<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Perform Enrichment analysis on differentially expressed genes or phospho-sites — enrichDifferential • SmartPhos</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform Enrichment analysis on differentially expressed genes or phospho-sites — enrichDifferential"><meta name="description" content="enrichDifferential performs enrichment analysis on differentially
expressed genes and phospho-sites for either pathway or phospho-specific
enrichment, depending on the input parameters. It supports multiple
statistical methods such as PAGE and GSEA for pathway enrichment and a
Kolmogorov-Smirnov approach for phospho-enrichment."><meta property="og:description" content="enrichDifferential performs enrichment analysis on differentially
expressed genes and phospho-sites for either pathway or phospho-specific
enrichment, depending on the input parameters. It supports multiple
statistical methods such as PAGE and GSEA for pathway enrichment and a
Kolmogorov-Smirnov approach for phospho-enrichment."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SmartPhos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.25</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SmartPhos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/SmartPhos_Shiny.html">SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Bioconductor/SmartPhos/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Perform Enrichment analysis on differentially expressed genes or phospho-sites</h1>
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/SmartPhos/blob/main/R/enrichment.R" class="external-link"><code>R/enrichment.R</code></a></small>
      <div class="d-none name"><code>enrichDifferential.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>enrichDifferential</code> performs enrichment analysis on differentially
expressed genes and phospho-sites for either pathway or phospho-specific
enrichment, depending on the input parameters. It supports multiple
statistical methods such as PAGE and GSEA for pathway enrichment and a
Kolmogorov-Smirnov approach for phospho-enrichment.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">enrichDifferential</span><span class="op">(</span></span>
<span>  <span class="va">dea</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Pathway enrichment"</span>, <span class="st">"Phospho-signature enrichment"</span><span class="op">)</span>,</span>
<span>  gsaMethod <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PAGE"</span>, <span class="st">"GSEA"</span><span class="op">)</span>,</span>
<span>  <span class="va">geneSet</span>,</span>
<span>  <span class="va">ptmSet</span>,</span>
<span>  statType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"stat"</span>, <span class="st">"log2FC"</span><span class="op">)</span>,</span>
<span>  nPerm <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  sigLevel <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  ifFDR <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dea">dea<a class="anchor" aria-label="anchor" href="#arg-dea"></a></dt>
<dd><p>A <code>data frame</code> containing the differential expression
analysis results. It should include columns like
  `pvalue`, `Gene` (or `site`), `stat`, and `log2FC`.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>A <code>character</code> string indicating the type of enrichment.
Options are `"Pathway enrichment"` or `"Phospho-signature enrichment"`.</p></dd>


<dt id="arg-gsamethod">gsaMethod<a class="anchor" aria-label="anchor" href="#arg-gsamethod"></a></dt>
<dd><p>A <code>character</code> string specifying the gene set analysis
method for pathway enrichment. Options are `"PAGE"` or `"GSEA"`.</p></dd>


<dt id="arg-geneset">geneSet<a class="anchor" aria-label="anchor" href="#arg-geneset"></a></dt>
<dd><p>A gene set collection to use for pathway enrichment.</p></dd>


<dt id="arg-ptmset">ptmSet<a class="anchor" aria-label="anchor" href="#arg-ptmset"></a></dt>
<dd><p>A post-translational modification (PTM) set database for
phospho-enrichment analysis.</p></dd>


<dt id="arg-stattype">statType<a class="anchor" aria-label="anchor" href="#arg-stattype"></a></dt>
<dd><p>A <code>character</code> string specifying the statistic type to
use. Options are `"stat"` or `"log2FC"`.</p></dd>


<dt id="arg-nperm">nPerm<a class="anchor" aria-label="anchor" href="#arg-nperm"></a></dt>
<dd><p>A <code>numeric</code> specifying the number of permutations for GSEA.
Default is 100.</p></dd>


<dt id="arg-siglevel">sigLevel<a class="anchor" aria-label="anchor" href="#arg-siglevel"></a></dt>
<dd><p>A <code>numeric</code> value representing the significance
threshold for filtering results. Ddefault is 0.05.</p></dd>


<dt id="arg-iffdr">ifFDR<a class="anchor" aria-label="anchor" href="#arg-iffdr"></a></dt>
<dd><p>A <code>logical</code> value indicating whether to filter results
using FDR-adjusted p-values. Default is `FALSE`.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame containing the results of the enrichment analysis,
including columns such as the gene set name,
  statistical significance, and adjusted p-values.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The `enrichDifferential` function performs either pathway enrichment or
phospho-enrichment analysis based on the `type` parameter.
For pathway enrichment, it uses either the PAGE or GSEA method with a
provided gene set collection. For phospho-enrichment,
it uses a Kolmogorov-Smirnov test with a PTM set database. Results can be
filtered by significance level and optionally adjusted for FDR.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.sysbio.se/piano" class="external-link">piano</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Load multiAssayExperiment object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dia_example"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Get SummarizedExperiment object</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">dia_example</span><span class="op">[[</span><span class="st">"Phosphoproteome"</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">dia_example</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Preprocess the proteome assay</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="preprocessPhos.html">preprocessPhos</a></span><span class="op">(</span><span class="va">se</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Call the function to perform differential expression analyis</span></span></span>
<span class="r-in"><span><span class="va">dea</span> <span class="op">&lt;-</span> <span class="fu"><a href="performDifferentialExp.html">performDifferentialExp</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">result</span>, assay <span class="op">=</span> <span class="st">"Intensity"</span>,</span></span>
<span class="r-in"><span>method <span class="op">=</span> <span class="st">"limma"</span>, reference <span class="op">=</span> <span class="st">"1stCrtl"</span>, target <span class="op">=</span> <span class="st">"EGF"</span>,</span></span>
<span class="r-in"><span>condition <span class="op">=</span> <span class="st">"treatment"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Load the gene set</span></span></span>
<span class="r-in"><span><span class="va">genesetPath</span> <span class="op">&lt;-</span> <span class="va">appDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"shiny-app/geneset"</span>,</span></span>
<span class="r-in"><span>package <span class="op">=</span> <span class="st">"SmartPhos"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">inGMT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/piano/man/loadGSC.html" class="external-link">loadGSC</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">genesetPath</span>,<span class="st">"/Cancer_Hallmark.gmt"</span><span class="op">)</span>,type<span class="op">=</span><span class="st">"gmt"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Call the function</span></span></span>
<span class="r-in"><span><span class="va">resTab</span> <span class="op">&lt;-</span> <span class="fu">enrichDifferential</span><span class="op">(</span>dea <span class="op">=</span> <span class="va">dea</span><span class="op">$</span><span class="va">resDE</span>, type <span class="op">=</span> <span class="st">"Pathway enrichment"</span>,</span></span>
<span class="r-in"><span>gsaMethod <span class="op">=</span> <span class="st">"PAGE"</span>, geneSet <span class="op">=</span> <span class="va">inGMT</span>, statType <span class="op">=</span> <span class="st">"stat"</span>, nPerm <span class="op">=</span> <span class="fl">200</span>,</span></span>
<span class="r-in"><span>sigLevel <span class="op">=</span> <span class="fl">0.05</span>, ifFDR <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running gene set analysis:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Checking arguments...done!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Final gene/gene-set association: 20 genes and 22 gene sets</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Details:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Calculating gene set statistics from 20 out of 65 gene-level statistics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Removed 4366 genes from GSC due to lack of matching gene statistics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Removed 28 gene sets containing no genes after gene removal</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Removed additionally 0 gene sets not matching the size limits</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Loaded additional information for 22 gene sets</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Calculating gene set statistics...done!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Calculating gene set significance...done!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Adjusting for multiple testing...done!</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Shubham Agrawal, Junyan Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

