<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Perform Differential Expression Analysis — performDifferentialExp • SmartPhos</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform Differential Expression Analysis — performDifferentialExp"><meta name="description" content="performDifferentialExp performs differential expression analysis on a
given SummarizedExperiment object using either the limma or
ProDA method."><meta property="og:description" content="performDifferentialExp performs differential expression analysis on a
given SummarizedExperiment object using either the limma or
ProDA method."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SmartPhos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.28</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SmartPhos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/SmartPhos_Shiny.html">SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Bioconductor/SmartPhos/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Perform Differential Expression Analysis</h1>
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/SmartPhos/blob/main/R/differential_expression.R" class="external-link"><code>R/differential_expression.R</code></a></small>
      <div class="d-none name"><code>performDifferentialExp.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>performDifferentialExp</code> performs differential expression analysis on a
given <code>SummarizedExperiment</code> object using either the <code>limma</code> or
<code>ProDA</code> method.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">performDifferentialExp</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  <span class="va">assay</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"limma"</span>, <span class="st">"ProDA"</span><span class="op">)</span>,</span>
<span>  condition <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">reference</span>,</span>
<span>  <span class="va">target</span>,</span>
<span>  refTime <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  targetTime <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pairedTtest <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-se">se<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>A <code>SummarizedExperiment</code> object containing the data.</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>A <code>character</code> string specifying the assay to use for the
analysis.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A <code>character</code> string specifying the method to use for
differential expression analysis ("limma" or "ProDA"). Default is "limma".</p></dd>


<dt id="arg-condition">condition<a class="anchor" aria-label="anchor" href="#arg-condition"></a></dt>
<dd><p>A <code>character</code> string specifying the condition column in
colData(se). Default is <code>NULL</code>.</p></dd>


<dt id="arg-reference">reference<a class="anchor" aria-label="anchor" href="#arg-reference"></a></dt>
<dd><p>A <code>character</code> string or vector specifying the reference
group.</p></dd>


<dt id="arg-target">target<a class="anchor" aria-label="anchor" href="#arg-target"></a></dt>
<dd><p>A <code>character</code> string or vector specifying the target
group.</p></dd>


<dt id="arg-reftime">refTime<a class="anchor" aria-label="anchor" href="#arg-reftime"></a></dt>
<dd><p>A <code>character</code> string or vector specifying the reference
time points. Default is <code>NULL</code>.</p></dd>


<dt id="arg-targettime">targetTime<a class="anchor" aria-label="anchor" href="#arg-targettime"></a></dt>
<dd><p>A <code>character</code> string or vector specifying the target
time points. Default is <code>NULL</code>.</p></dd>


<dt id="arg-pairedttest">pairedTtest<a class="anchor" aria-label="anchor" href="#arg-pairedttest"></a></dt>
<dd><p>A <code>logical</code> value specifying to perform paired t-test
or not. Default is <code>FALSE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing:</p>
<dl><dt>resDE</dt>
<dd><p>A <code>tibble</code> with the differential expression results.</p></dd>

<dt>seSub</dt>
<dd><p>A <code>SummarizedExperiment</code> object subset to the samples used
in the analysis.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function is designed to facilitate differential expression analysis on
a <code>SummarizedExperiment</code> (SE) object. The function allows users to
specify various parameters to tailor the analysis to their specific
experimental setup.</p>
<p>The main steps of the function are as follows:</p>
<p>1. Sample Selection: Based on the provided <code>condition</code>,
<code>reference</code>, and <code>target</code> arguments, the function identifies the
relevant samples for the analysis. If time points (<code>refTime</code> and
<code>targetTime</code>) are provided, it further refines the sample selection.</p>
<p>2. Subsetting the SE Object: The SE object is subsetted to include only the
selected samples. A new column <code>comparison</code> is added to the
<code>colData</code>, indicating whether each sample belongs to the reference or
target group.</p>
<p>3. Design Matrix Construction: The function constructs a design matrix for
the differential expression analysis. If the SE object contains a
<code>subjectID</code> column, this is included in the design to account for
repeated measures or paired samples.</p>
<p>4. Differential Expression Analysis: Depending on the specified
<code>method</code>, the function performs the differential expression analysis
using either the <code>limma</code> or <code>ProDA</code> package:
    - <code>Limma</code>: The function fits a linear model to the expression data
    and applies empirical Bayes moderation to the standard errors. The
    results are then extracted and formatted.
    - <code>ProDA</code>: The function fits a probabilistic dropout model to the
    expression data and tests for differential expression. The results are
    then extracted and formatted.</p>
<p>5. Result Formatting: The differential expression results are merged with
the metadata from the SE object, and the resulting table is formatted into a
tibble. The table includes columns for log2 fold change (log2FC), test
statistic (stat), p-value (pvalue), adjusted p-value (padj), and gene/feature
ID (ID).</p>
<p>The function returns a <code>list</code> containing the formatted differential
expression results and the subsetted SE object. This allows users to further
explore or visualize the results as needed.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Load multiAssayExperiment object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dda_example"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Get SummarizedExperiment object</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">dda_example</span><span class="op">[[</span><span class="st">"Proteome"</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">dda_example</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Preprocess the proteome assay</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="preprocessProteome.html">preprocessProteome</a></span><span class="op">(</span><span class="va">se</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Call the function to perform differential expression analyis</span></span></span>
<span class="r-in"><span><span class="fu">performDifferentialExp</span><span class="op">(</span>se <span class="op">=</span> <span class="va">result</span>, assay <span class="op">=</span> <span class="st">"Intensity"</span>, method <span class="op">=</span> <span class="st">"limma"</span>,</span></span>
<span class="r-in"><span>reference <span class="op">=</span> <span class="st">"1stCrtl"</span>, target <span class="op">=</span> <span class="st">"EGF"</span>, condition <span class="op">=</span> <span class="st">"treatment"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $resDE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 83 × 8</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ID    log2FC  stat  pvalue  padj UniprotID            Gene     PeptideCounts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> p99     2.02  4.75 0.001<span style="text-decoration: underline;">53</span> 0.127 O00743               PPP6C    12           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> p95    -<span style="color: #BB0000;">2.73</span> -<span style="color: #BB0000;">4.03</span> 0.003<span style="text-decoration: underline;">96</span> 0.164 O00571               DDX3X    45           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> p48     2.53  3.73 0.006<span style="text-decoration: underline;">00</span> 0.166 E9PRG8               C11orf98 4            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> p93     2.46  3.04 0.016<span style="text-decoration: underline;">5</span>  0.276 O00560               SDCBP    10           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> p54    -<span style="color: #BB0000;">2.23</span> -<span style="color: #BB0000;">3.03</span> 0.016<span style="text-decoration: underline;">6</span>  0.276 O00151               PDLIM1   24           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> p38     2.83  2.43 0.041<span style="text-decoration: underline;">6</span>  0.549 CON__P48668;P48668   KRT6C    46;46        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> p96     1.64  2.22 0.057<span style="text-decoration: underline;">7</span>  0.549 O00622               CYR61    15           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> p100   -<span style="color: #BB0000;">1.40</span> -<span style="color: #BB0000;">2.21</span> 0.058<span style="text-decoration: underline;">5</span>  0.549 O00764               PDXK     19           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> p47    -<span style="color: #BB0000;">2.03</span> -<span style="color: #BB0000;">2.14</span> 0.065<span style="text-decoration: underline;">5</span>  0.549 E9PAV3;Q13765;Q9BZK3 NACA     9;9;2        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> p71    -<span style="color: #BB0000;">1.55</span> -<span style="color: #BB0000;">2.08</span> 0.072<span style="text-decoration: underline;">0</span>  0.549 O00299               CLIC1    15           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 73 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $seSub</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 83 8 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): Intensity</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(83): p2 p3 ... p99 p100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(3): UniprotID Gene PeptideCounts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(8): Full_1stCrtl_0min_rep1 Full_1stCrtl_0min_rep2 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Full_EGF_40min_rep1 Full_EGF_40min_rep2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(7): sample sampleType ... batch comparison</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Shubham Agrawal, Junyan Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

