<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot Heatmap of Intensity assay — plotHeatmap • SmartPhos</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot Heatmap of Intensity assay — plotHeatmap"><meta name="description" content="plotHeatmap generates a heatmap for intensity assay for different
conditions, including top variants, differentially expressed genes, and
selected time series clusters."><meta property="og:description" content="plotHeatmap generates a heatmap for intensity assay for different
conditions, including top variants, differentially expressed genes, and
selected time series clusters."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SmartPhos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.25</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SmartPhos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/SmartPhos_Shiny.html">SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Bioconductor/SmartPhos/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot Heatmap of Intensity assay</h1>
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/SmartPhos/blob/main/R/basic_visualizations.R" class="external-link"><code>R/basic_visualizations.R</code></a></small>
      <div class="d-none name"><code>plotHeatmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>plotHeatmap</code> generates a heatmap for intensity assay for different
conditions, including top variants, differentially expressed genes, and
selected time series clusters.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotHeatmap</span><span class="op">(</span></span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Top variant"</span>, <span class="st">"Differentially expressed"</span>, <span class="st">"Selected time series cluster"</span><span class="op">)</span>,</span>
<span>  <span class="va">se</span>,</span>
<span>  data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  top <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  cutCol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  cutRow <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  clustCol <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  clustRow <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  annotationCol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>A <code>character</code> string specifying the type of heatmap to plot.
Options are "Top variant", "Differentially expressed", and "Selected time
series cluster".</p></dd>


<dt id="arg-se">se<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>A <code>SummarizedExperiment</code> object containing the imputed
intensity assay.</p></dd>


<dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>An optional <code>data frame</code> containing additional data for
"Differentially expressed" and "Selected time series cluster" types. Default
is <code>NULL</code>.</p></dd>


<dt id="arg-top">top<a class="anchor" aria-label="anchor" href="#arg-top"></a></dt>
<dd><p>A <code>numeric</code> value specifying the number of top variants to
plot. Default is 100.</p></dd>


<dt id="arg-cutcol">cutCol<a class="anchor" aria-label="anchor" href="#arg-cutcol"></a></dt>
<dd><p>A <code>numeric</code> value specifying the number of clusters for
columns. Default is 1.</p></dd>


<dt id="arg-cutrow">cutRow<a class="anchor" aria-label="anchor" href="#arg-cutrow"></a></dt>
<dd><p>A <code>numeric</code> value specifying the number of clusters for
rows. Default is 1.</p></dd>


<dt id="arg-clustcol">clustCol<a class="anchor" aria-label="anchor" href="#arg-clustcol"></a></dt>
<dd><p>A <code>logical</code> value indicating whether to cluster columns.
Default is <code>TRUE</code>.</p></dd>


<dt id="arg-clustrow">clustRow<a class="anchor" aria-label="anchor" href="#arg-clustrow"></a></dt>
<dd><p>A <code>logical</code> value indicating whether to cluster rows.
Default is <code>TRUE</code>.</p></dd>


<dt id="arg-annotationcol">annotationCol<a class="anchor" aria-label="anchor" href="#arg-annotationcol"></a></dt>
<dd><p>A <code>character</code> vector specifying the columns in the
metadata to use for annotation. Default is <code>NULL</code>.</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>A <code>character</code> string specifying the title of the heatmap.
Default is <code>NULL</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>pheatmap</code> object showing the heatmap of Intensity data.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function creates a heatmap using the Intensity assay from a
<code>SummarizedExperiment</code> object. The heatmap can show the top variants
based on standard deviation, differentially expressed genes, or selected time
series clusters. Row normalization is performed, and the heatmap can include
annotations based on specified metadata columns.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Load multiAssayExperiment object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dia_example"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Get SummarizedExperiment object</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">dia_example</span><span class="op">[[</span><span class="st">"Phosphoproteome"</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">dia_example</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Generate the imputed assay</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="preprocessPhos.html">preprocessPhos</a></span><span class="op">(</span>seData <span class="op">=</span> <span class="va">se</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span>, impute <span class="op">=</span> <span class="st">"QRILC"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing along margin 2 (samples/columns).</span>
<span class="r-in"><span><span class="co"># Plot heatmap for top variant</span></span></span>
<span class="r-in"><span><span class="fu">plotHeatmap</span><span class="op">(</span>type <span class="op">=</span> <span class="st">"Top variant"</span>, top <span class="op">=</span> <span class="fl">10</span>, se <span class="op">=</span> <span class="va">result</span>, cutCol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotHeatmap-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Shubham Agrawal, Junyan Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

