<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Perform Fisher's Exact Test on Gene Sets — runFisher • SmartPhos</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform Fisher's Exact Test on Gene Sets — runFisher"><meta name="description" content="runFisher performs Fisher's Exact Test to determine the enrichment of
a set of genes within reference gene sets."><meta property="og:description" content="runFisher performs Fisher's Exact Test to determine the enrichment of
a set of genes within reference gene sets."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SmartPhos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.26</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SmartPhos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/SmartPhos_Shiny.html">SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Bioconductor/SmartPhos/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Perform Fisher's Exact Test on Gene Sets</h1>
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/SmartPhos/blob/main/R/enrichment.R" class="external-link"><code>R/enrichment.R</code></a></small>
      <div class="d-none name"><code>runFisher.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>runFisher</code> performs Fisher's Exact Test to determine the enrichment of
a set of genes within reference gene sets.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runFisher</span><span class="op">(</span><span class="va">genes</span>, <span class="va">reference</span>, <span class="va">inputSet</span>, ptm <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>A <code>character</code> vector of genes of interest.</p></dd>


<dt id="arg-reference">reference<a class="anchor" aria-label="anchor" href="#arg-reference"></a></dt>
<dd><p>A <code>character</code> vector of reference genes.</p></dd>


<dt id="arg-inputset">inputSet<a class="anchor" aria-label="anchor" href="#arg-inputset"></a></dt>
<dd><p>A <code>list</code> containing gene set collections. If ptm is
<code>TRUE</code>, this should be a data frame with specific columns.</p></dd>


<dt id="arg-ptm">ptm<a class="anchor" aria-label="anchor" href="#arg-ptm"></a></dt>
<dd><p><code>Logical</code>. If <code>TRUE</code>, perform the test on
post-translational modification (PTM) gene sets. Default is <code>FALSE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>data frame</code> with the results of the Fisher's Exact Test,
including the gene set name, the number of genes in the set, set size,
p-value, adjusted p-value, and the genes in the set.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function can operate in two modes: standard gene sets and PTM-specific
gene sets. For PTM-specific gene sets, additional filtering and processing
are performed.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.sysbio.se/piano" class="external-link">piano</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Load multiAssayExperiment object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dda_example"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Get SummarizedExperiment object</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">dda_example</span><span class="op">[[</span><span class="st">"Proteome"</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">dda_example</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Preprocess the proteome assay</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="preprocessProteome.html">preprocessProteome</a></span><span class="op">(</span><span class="va">se</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Call the function to perform differential expression analyis</span></span></span>
<span class="r-in"><span><span class="va">de</span> <span class="op">&lt;-</span> <span class="fu"><a href="performDifferentialExp.html">performDifferentialExp</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">result</span>, assay <span class="op">=</span> <span class="st">"Intensity"</span>,</span></span>
<span class="r-in"><span>method <span class="op">=</span> <span class="st">"limma"</span>, reference <span class="op">=</span> <span class="st">"1stCrtl"</span>, target <span class="op">=</span> <span class="st">"EGF"</span>,</span></span>
<span class="r-in"><span>condition <span class="op">=</span> <span class="st">"treatment"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genesList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">de</span><span class="op">$</span><span class="va">resDE</span><span class="op">$</span><span class="va">Gene</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">referenceList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span><span class="op">$</span><span class="va">Gene</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genesetPath</span> <span class="op">&lt;-</span> <span class="va">appDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"shiny-app/geneset"</span>,</span></span>
<span class="r-in"><span>package <span class="op">=</span> <span class="st">"SmartPhos"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">inGMT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/piano/man/loadGSC.html" class="external-link">loadGSC</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">genesetPath</span>,<span class="st">"/Cancer_Hallmark.gmt"</span><span class="op">)</span>,type<span class="op">=</span><span class="st">"gmt"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Run the function</span></span></span>
<span class="r-in"><span><span class="fu">runFisher</span><span class="op">(</span>genes <span class="op">=</span> <span class="va">genesList</span>, reference <span class="op">=</span> <span class="va">referenceList</span>, inputSet <span class="op">=</span> <span class="va">inGMT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                          Name Gene.number Set.size pval</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            HALLMARK_TNFA_SIGNALING_VIA_NFKB           1        1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                            HALLMARK_HYPOXIA           3        3    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                    HALLMARK_MITOTIC_SPINDLE           3        3    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            HALLMARK_IL6_JAK_STAT3_SIGNALING           1        1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                         HALLMARK_DNA_REPAIR           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                     HALLMARK_G2M_CHECKPOINT           1        1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7                          HALLMARK_APOPTOSIS           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8            HALLMARK_ESTROGEN_RESPONSE_EARLY           5        5    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9             HALLMARK_ESTROGEN_RESPONSE_LATE           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10                 HALLMARK_ANDROGEN_RESPONSE           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11                        HALLMARK_MYOGENESIS           3        3    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12                 HALLMARK_PROTEIN_SECRETION           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13                   HALLMARK_APICAL_JUNCTION           1        1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14                        HALLMARK_COMPLEMENT           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15         HALLMARK_UNFOLDED_PROTEIN_RESPONSE           1        1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16                  HALLMARK_MTORC1_SIGNALING           4        4    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17                       HALLMARK_E2F_TARGETS           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18                    HALLMARK_MYC_TARGETS_V1           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19                    HALLMARK_MYC_TARGETS_V2           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20 HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21             HALLMARK_XENOBIOTIC_METABOLISM           1        1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 22         HALLMARK_OXIDATIVE_PHOSPHORYLATION           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23                        HALLMARK_GLYCOLYSIS           4        4    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24   HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25                    HALLMARK_UV_RESPONSE_DN           1        1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26                   HALLMARK_HEME_METABOLISM           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27               HALLMARK_IL2_STAT5_SIGNALING           1        1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28              HALLMARK_BILE_ACID_METABOLISM           1        1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 29                        HALLMARK_PEROXISOME           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 30                 HALLMARK_KRAS_SIGNALING_DN           2        2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           Genes padj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1         CYR61    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  CYR61, I....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  ARHGAP10....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          HAX1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  SDCBP, S....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        DDX39A    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7   DFFA, DNM1L    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  KRT19, K....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9  KRT19, S....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10  KRT19, KRT8    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 BIN1, MY....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 AP3B1, DNM1L    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13         WASL    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14 ATOX1, KIF2A    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15        NOP56    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16 PLOD2, P....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17 NOP56, D....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18 NOP56, P....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19  PES1, NOP56    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20 CYR61, PLOD2    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21       PGRMC1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 22 PDHX, NDUFS8    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23 PLOD2, K....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24 PDLIM1, ....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25        CYR61    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26 SDCBP, PSMD9    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27        MYO1C    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28      SULT2B1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 29 NUDT19, ....    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 30  KRT15, KRT5    1</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Shubham Agrawal, Junyan Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

