<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A computational framework for phosphoproteomics data analysis with an interactive ShinyApp â€¢ SmartPhos</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A computational framework for phosphoproteomics data analysis with an interactive ShinyApp">
<meta name="description" content="Phosphoproteomics provides rich information for dissecting pathway activities and therefore is becoming vital in basic and translational biomedical research. However, processing and analyzing phosphoproteomics data are still daunting tasks, especially for researchers without coding experience. To facilitate and streamline phosphoproteomics data analysis, we developed SmartPhos, an R package for the pre-processing, quality control, and exploratory analysis of phosphoproteomics data generated by mass-spectrometry. SmartPhos can process outputs from MaxQuant and Spectronaut either using the R command line or in an interactive ShinyApp called SmartPhos Explorer. Besides commonly used preprocessing steps, such as normalization, transformation, imputation, and batch effect correction, our framework features a novel method for correcting normalization artifacts observed in phosphoproteomics, especially when large global phosphorylation changes are expected, by taking both phospho-enriched and unenriched samples into account. In addition, the SmartPhos Explorer ShinyApp included in our R package provides a user-friendly and interactive one-stop solution for performing exploratory data analysis (PCA, hierarchical clustering, etc.), differential expression, time-series clustering, gene set enrichment analysis, and kinase activity analysis easily without the knowledge of coding or the underlying statistical model. To ensure reproducibility in users' analyses, all user inputs and interactions are documented and can be downloaded as a log file. Our computational framework can help unleash the full potential of phosphoproteomic data in biomedical research by lowering the barrier of data analysis, especially for large data sets with complex designs, as well as by promoting reproducible research.">
<meta property="og:description" content="Phosphoproteomics provides rich information for dissecting pathway activities and therefore is becoming vital in basic and translational biomedical research. However, processing and analyzing phosphoproteomics data are still daunting tasks, especially for researchers without coding experience. To facilitate and streamline phosphoproteomics data analysis, we developed SmartPhos, an R package for the pre-processing, quality control, and exploratory analysis of phosphoproteomics data generated by mass-spectrometry. SmartPhos can process outputs from MaxQuant and Spectronaut either using the R command line or in an interactive ShinyApp called SmartPhos Explorer. Besides commonly used preprocessing steps, such as normalization, transformation, imputation, and batch effect correction, our framework features a novel method for correcting normalization artifacts observed in phosphoproteomics, especially when large global phosphorylation changes are expected, by taking both phospho-enriched and unenriched samples into account. In addition, the SmartPhos Explorer ShinyApp included in our R package provides a user-friendly and interactive one-stop solution for performing exploratory data analysis (PCA, hierarchical clustering, etc.), differential expression, time-series clustering, gene set enrichment analysis, and kinase activity analysis easily without the knowledge of coding or the underlying statistical model. To ensure reproducibility in users' analyses, all user inputs and interactions are documented and can be downloaded as a log file. Our computational framework can help unleash the full potential of phosphoproteomic data in biomedical research by lowering the barrier of data analysis, especially for large data sets with complex designs, as well as by promoting reproducible research.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">SmartPhos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/SmartPhos.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/SmartPhos_Shiny.html">SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p>This repository contains the source code for SmartPhos package, which provides a one-stop solution for the processing and analysis of mass-spectrometry-based phopho-proteomics data. The package is still under active development currently.<br>
For users with coding experience, they can use the functions provided by SmartPhos to perform data preprocessing. For users without coding experience, SmartPhos also provides a Shiny app for the interactive analysis of proteomic and phosphoproteomic data.<br>
- User guide: <a href="https://lu-group-ukhd.github.io/SmartPhos/vignette.html" class="uri">https://lu-group-ukhd.github.io/SmartPhos/vignette.html</a></p>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SmartPhos</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Shubham Agrawal <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0009-0005-2630-9342" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Junyan Lu <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-9211-0746" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Shubham Agrawal, Junyan Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
