<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SmartPhos: a pipline for processing and analysis of phosphoproteome data • SmartPhos</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SmartPhos: a pipline for processing and analysis of phosphoproteome data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SmartPhos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/SmartPhos.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/SmartPhos_Shiny.html">SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>SmartPhos: a pipline for processing and analysis of phosphoproteome data</h1>
                        <h4 data-toc-skip class="author">Shubham
Agrawal, Junyan Lu</h4>
            
            <h4 data-toc-skip class="date">2024-12-10</h4>
      

      <div class="d-none name"><code>SmartPhos.Rmd</code></div>
    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Phosphoproteomics provides rich information for dissecting pathway
activities and therefore is becoming vital in basic and translational
biomedical research. To facilitate and streamline phosphoproteomics data
analysis, we developed <code>SmartPhos</code>, an R package for the
pre-processing, quality control, and exploratory analysis of
phosphoproteomics data generated by mass-spectrometry. SmartPhos can
process outputs from <code>MaxQuant</code> and <code>Spectronaut</code>
either using the R command line or in an interactive ShinyApp called
<code>SmartPhos Explorer</code>. Besides commonly used preprocessing
steps, such as normalization, transformation, imputation, and batch
effect correction, our package features a novel method for correcting
normalization artifacts observed in phosphoproteomics, especially when
large global phosphorylation changes are expected, by taking both
phospho-enriched and unenriched samples into account.</p>
<p>In addition, the <code>SmartPhos Explorer</code>
<code>ShinyApp</code> included in our R package provides a user-friendly
and interactive one-stop solution for performing exploratory data
analysis (PCA, hierarchical clustering, etc.), differential expression,
time-series clustering, gene set enrichment analysis, and kinase
activity analysis easily without the knowledge of coding or the
underlying statistical model.</p>
<p>This vignette focuses on using <code>SmartPhos</code> with the R
command line.</p>
</div>
<div class="section level2">
<h2 id="install-and-load-smartphos-package">Install and load SmartPhos Package<a class="anchor" aria-label="anchor" href="#install-and-load-smartphos-package"></a>
</h2>
<p>To install <code>SmartPhos</code> enter the following to the
<code>R</code> console</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SmartPhos"</span><span class="op">)</span></span></code></pre></div>
<p>Load the package</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lu-group-ukhd.github.io/SmartPhos/">SmartPhos</a></span><span class="op">)</span></span></code></pre></div>
<p>Load the R object</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dia_example"</span><span class="op">)</span></span>
<span><span class="va">dia_example</span></span></code></pre></div>
<pre><code><span><span class="co">## A MultiAssayExperiment object of 2 listed</span></span>
<span><span class="co">##  experiments with user-defined names and respective classes.</span></span>
<span><span class="co">##  Containing an ExperimentList class object of length 2:</span></span>
<span><span class="co">##  [1] Phosphoproteome: SummarizedExperiment with 500 rows and 47 columns</span></span>
<span><span class="co">##  [2] Proteome: SummarizedExperiment with 500 rows and 47 columns</span></span>
<span><span class="co">## Functionality:</span></span>
<span><span class="co">##  experiments() - obtain the ExperimentList instance</span></span>
<span><span class="co">##  colData() - the primary/phenotype DataFrame</span></span>
<span><span class="co">##  sampleMap() - the sample coordination DataFrame</span></span>
<span><span class="co">##  `$`, `[`, `[[` - extract colData columns, subset, or experiment</span></span>
<span><span class="co">##  *Format() - convert into a long or wide DataFrame</span></span>
<span><span class="co">##  assays() - convert ExperimentList to a SimpleList of matrices</span></span>
<span><span class="co">##  exportClass() - save data to flat files</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="preprocessing-the-assay-basic-visualization-pca-and-heatmaps">Preprocessing the assay, basic visualization, PCA and Heatmaps<a class="anchor" aria-label="anchor" href="#preprocessing-the-assay-basic-visualization-pca-and-heatmaps"></a>
</h2>
<p>Extract the SummarizedExperiment object from the multiAssayExperiment
object</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">dia_example</span><span class="op">[[</span><span class="st">"Phosphoproteome"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">dia_example</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SummarizedExperiment </span></span>
<span><span class="co">## dim: 500 47 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): Intensity</span></span>
<span><span class="co">## rownames(500): s1 s2 ... s499 s500</span></span>
<span><span class="co">## rowData names(7): UniprotID Gene ... Sequence site</span></span>
<span><span class="co">## colnames(47): FullProteome_1stCrtl_0min_rep2</span></span>
<span><span class="co">##   FullProteome_1stCrtl_0min_rep3 ... Phospho_HGF_24h_rep1</span></span>
<span><span class="co">##   Phospho_HGF_100min_rep1</span></span>
<span><span class="co">## colData names(6): sample treatment ... sampleType sampleName</span></span></code></pre>
<div class="section level3">
<h3 id="preprocessing-options">Preprocessing options<a class="anchor" aria-label="anchor" href="#preprocessing-options"></a>
</h3>
<p>SmartPhos package performs some pre-filtering and preprocessing based
on various threshold values during the generation of the
multiAssayExperiment object. On top of that, this panel has different
options for preprocessing of the selected assay. The options provided by
the shiny app are as follows:</p>
<ul>
<li>
<strong>Normalization correction</strong>: This option is present
only if the phophoproteomics data is present. It allows for the
correction of normalization artefacts introduced by the
Spectronaut.</li>
<li>
<strong>Normalize</strong> phospho intensity by the corresponding
protein expression.</li>
<li>
<strong>Transformation</strong>: The transformation methods
available are: log2 and vst (variance stabilizing transformation).</li>
<li>
<strong>Normalization</strong>: Normalization strategy depends on
the selected transformation method. Therefore, the user has option of
<em>Yes</em> or <em>No</em>. If the user selects <em>Yes</em>, then for
log2 and no transformation, <em>median scaling</em> is applied and for
vst, <em>vsn</em> is applied.</li>
<li>
<strong>Missing values</strong>: Users can select what percentage of
missing values are allowed. Proteins with missing values above the
selected threshold will be removed from all analyses.</li>
<li>
<strong>Imputation</strong>: Currently four imputation methods are
available:
<ul>
<li>QRILC (Quantile Regression Imputation of Left-Censored data)</li>
<li>MinDet (Deterministic minimal value approach)</li>
<li>BPCA (Bayesian PCA)</li>
<li>MLE (Maximum Likelihood Estimation)</li>
<li>Random forest</li>
</ul>
</li>
<li>
<strong>Batch effects removal</strong>: This option provides the
abilty to correct for batch effects. It uses
<em>removeBatchEffect()</em> from the <em>limma</em> package. The users
can select the maximum of two columns. The columns used for batch
effects removal should be present in the fileTable.txt file.</li>
</ul>
<p>The preprocessing (transformation, normalization, imputation, etc.)
on the intensity assay can be performed by running the following
command:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">newSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocessPhos.html">preprocessPhos</a></span><span class="op">(</span>seData <span class="op">=</span> <span class="va">se</span>, transform <span class="op">=</span> <span class="st">"log2"</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                        impute <span class="op">=</span> <span class="st">"QRILC"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="different-visualization-options">Different visualization options<a class="anchor" aria-label="anchor" href="#different-visualization-options"></a>
</h3>
<p>Plot the boxplot of assay intensities:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotIntensity.html">plotIntensity</a></span><span class="op">(</span><span class="va">newSE</span>, color <span class="op">=</span> <span class="st">"replicate"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SmartPhos_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>Plot the completeness (percentage of non-missing) for each
sample:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMissing.html">plotMissing</a></span><span class="op">(</span><span class="va">newSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SmartPhos_files/figure-html/unnamed-chunk-6-1.png" width="700">
This is performed on the original assay and not on the imputed
assay.</p>
<p>Perform principal component analysis (PCA) by running the following
command:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># perform PCA</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">assays</span><span class="op">(</span><span class="va">newSE</span><span class="op">)</span><span class="op">[[</span><span class="st">"imputed"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, scale. <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># call the plotting function</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotPCA.html">plotPCA</a></span><span class="op">(</span>pca <span class="op">=</span> <span class="va">pca</span>, se <span class="op">=</span> <span class="va">newSE</span>, color <span class="op">=</span> <span class="st">"replicate"</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="SmartPhos_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Heatmap can be plotted using the plotHeatmap() function of SmartPhos.
There are three different heatmaps possible based on the
<strong>type</strong> argument:</p>
<ul>
<li>
<strong>Top variant</strong>: This allows the users to plot the
genes with highest variance. Users can decide the number of top variants
genes to plot. This option performs clustering automatically. The user
also has the option to divide the columns and rows of the heatmap into
specific number of clusters using the <strong>cutCol</strong> and
<strong>cutRow</strong> arguments.</li>
<li>
<strong>Differentially expressed</strong>: Allows to plot the
heatmaps for differentially expressed genes. The differential expression
analysis is available in <em>Differential expression</em> tab and can be
performed using ProDA or Limma.</li>
<li>
<strong>Selected time series cluster</strong>: After performing the
time-series clustering in another tab, users have option to plot the
heatmap of the selected cluster.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"Top variant"</span>, <span class="va">newSE</span>, top <span class="op">=</span> <span class="fl">10</span>, </span>
<span>            annotationCol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"replicate"</span>, <span class="st">"treatment"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SmartPhos_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="differential-expression-analysis">Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h2>
<p>The goal of performing differential expression analysis is to
quantify the expression levels of genes between different experimental
conditions using statistical tests.</p>
<p>If the users want paired t-test on the patient IDs, cell lines, etc,
then the users must have <strong>subjectID</strong> as one of the column
in the <em>fileTable.txt</em> file with the relevant information. The
subjectID column should also be selected in the additional column
annotations before the generation of multiAssayExperiment object.</p>
<p>The <em>performDifferentialExp()</em> has two methods for performing
differential expression analysis:</p>
<ul>
<li>
<strong>limma</strong>: uses linear models.</li>
<li>
<strong>ProDA</strong>: uses probabilistic dropout model.</li>
</ul>
<p>Limma method needs an assay with non-missing data and is faster than
the proDA method. ProDA method can work on assay with missing values
also.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dea</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/performDifferentialExp.html">performDifferentialExp</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">newSE</span>, assay <span class="op">=</span> <span class="st">"imputed"</span>, method <span class="op">=</span> <span class="st">"limma"</span>, </span>
<span>                              condition <span class="op">=</span> <span class="st">"treatment"</span>, reference <span class="op">=</span> <span class="st">"EGF"</span>, </span>
<span>                              target <span class="op">=</span> <span class="st">"1stCrtl"</span><span class="op">)</span></span></code></pre></div>
<p>‘dea’ is a list object which contains two values: ‘resDE’ with
tabular differential expression analysis results and ‘seSub’ which is
the subsetted SummarizedExperiment object based on the selected
condition.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dea</span><span class="op">$</span><span class="va">seSub</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SummarizedExperiment </span></span>
<span><span class="co">## dim: 172 13 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): Intensity imputed</span></span>
<span><span class="co">## rownames(172): s4 s6 ... s499 s500</span></span>
<span><span class="co">## rowData names(7): UniprotID Gene ... Sequence site</span></span>
<span><span class="co">## colnames(13): Phospho_1stCrtl_0min_rep2 Phospho_1stCrtl_0min_rep3 ...</span></span>
<span><span class="co">##   Phospho_EGF_24h_rep1 Phospho_EGF_100min_rep1</span></span>
<span><span class="co">## colData names(7): sample treatment ... sampleName comparison</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dea</span><span class="op">$</span><span class="va">resDE</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 172 × 12</span></span></span>
<span><span class="co">##    ID       log2FC  stat   pvalue   padj UniprotID   Gene  Multiplicity Position</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;I&lt;chr&gt;&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> s447      2.18   4.83 0.000<span style="text-decoration: underline;">373</span> 0.064<span style="text-decoration: underline;">1</span> O00401      WASL             1 256     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> s156      2.31   3.68 0.002<span style="text-decoration: underline;">96</span>  0.243  A0A590UJ23… PALM3            2 436     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> s184     -<span style="color: #BB0000;">1.62</span>  -<span style="color: #BB0000;">3.49</span> 0.004<span style="text-decoration: underline;">23</span>  0.243  A6NKD9      CCDC…            2 251     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> s318      3.15   3.07 0.009<span style="text-decoration: underline;">33</span>  0.401  O00139      KIF2A            1 484     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> s28      -<span style="color: #BB0000;">2.47</span>  -<span style="color: #BB0000;">2.87</span> 0.013<span style="text-decoration: underline;">6</span>   0.421  A0MZ66      SHTN1            1 101     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> s4        1.01   2.83 0.014<span style="text-decoration: underline;">7</span>   0.421  A0AV96      RBM47            1 519     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> s203     -<span style="color: #BB0000;">4.26</span>  -<span style="color: #BB0000;">2.75</span> 0.017<span style="text-decoration: underline;">2</span>   0.422  A7KAX9      ARHG…            2 44      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> s381      1.98   2.60 0.022<span style="text-decoration: underline;">6</span>   0.486  O00231      PSMD…            1 23      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> s173      0.927  2.45 0.030<span style="text-decoration: underline;">2</span>   0.534  A6NFI3      ZNF3…            3 10      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> s157      1.65   2.39 0.033<span style="text-decoration: underline;">7</span>   0.534  A0A590UJ23… PALM3            2 439     </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 162 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: Residue &lt;chr&gt;, Sequence &lt;chr&gt;, site &lt;chr&gt;</span></span></span></code></pre>
<p>Plot the volcano plot for the obtained differential expression
analysis results by running the following command:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotVolcano.html">plotVolcano</a></span><span class="op">(</span><span class="va">dea</span><span class="op">$</span><span class="va">resDE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SmartPhos_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p><strong>pFilter</strong> and <strong>fcFilter</strong> arguments can
be used to change the thresholds for the upregulated and downregulated
genes being represented in the volcano plot.</p>
<p>Use plotBox() to plot the intensity data of a specified gene, with
optional subject-specific lines:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBox.html">plotBox</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">dea</span><span class="op">$</span><span class="va">seSub</span>, id <span class="op">=</span> <span class="st">'s447'</span>, symbol <span class="op">=</span> <span class="st">"WASL"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SmartPhos_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="time-series-clustering">Time series clustering<a class="anchor" aria-label="anchor" href="#time-series-clustering"></a>
</h2>
<p>Time series clustering is performed to group proteins/phosphopeptides
based on how their level changes over time. The <em>clusterTS()</em>
function employs <strong>fuzzy c-means clustering</strong> algorithm
which considers the time-resolved trend, but not the expression levels.
Thus, members of the same cluster would have a similar trend over time
(e.g., all increasing or all decreasing), though their expression level
can be different.</p>
<p>To use this function, users must have logitudinal data with
<strong>timepoint</strong> annotation in their data. The time points are
either unit-less numbers (e.g., 1, 2, 3) or are in hour and/or minute,
which must be typed as “h” and “min”, respectively (e.g., 1min, 2min,
3h). Please notice that mixing the two said options (e.g., 1, 2min, 3)
or using other unit for time (e.g., 1hour, 2minute, 3day) will likely
lead to wrong results.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># call addZeroTime function to add zero timepoint to EGF treatment</span></span>
<span><span class="va">newSEzero</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addZeroTime.html">addZeroTime</a></span><span class="op">(</span><span class="va">newSE</span>, condition <span class="op">=</span> <span class="st">"treatment"</span>, treat <span class="op">=</span> <span class="st">"EGF"</span>, </span>
<span>                         zeroTreat <span class="op">=</span> <span class="st">"1stCrtl"</span>, </span>
<span>                         timeRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10min"</span>,<span class="st">"100min"</span>, <span class="st">"24h"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># extract the assay</span></span>
<span><span class="va">exprMat</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">newSEzero</span><span class="op">)</span></span>
<span><span class="co"># call the clustering function</span></span>
<span><span class="va">tsc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterTS.html">clusterTS</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">exprMat</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>‘tsc’ is a list object which contains two values: ‘cluster’ with
tabular time series clustering results and ‘plot’ with the clusters
plots.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tsc</span><span class="op">$</span><span class="va">cluster</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 580 × 7</span></span></span>
<span><span class="co">##    feature time                    value cluster   prob  cNum clusterNum   </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> s499    Phospho_EGF_10min_rep2  14.1  cluster1 0.402    10 cluster1 (10)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> s499    Phospho_EGF_10min_rep3  14.1  cluster1 0.402    10 cluster1 (10)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> s499    Phospho_EGF_100min_rep2 14.3  cluster1 0.402    10 cluster1 (10)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> s499    Phospho_EGF_100min_rep3 12.5  cluster1 0.402    10 cluster1 (10)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> s499    Phospho_EGF_10min_rep1   8.96 cluster1 0.402    10 cluster1 (10)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> s499    Phospho_EGF_24h_rep1    14.5  cluster1 0.402    10 cluster1 (10)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> s499    Phospho_EGF_100min_rep1 14.6  cluster1 0.402    10 cluster1 (10)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> s499    Phospho_EGF_0min_rep2   14.2  cluster1 0.402    10 cluster1 (10)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> s499    Phospho_EGF_0min_rep3   14.2  cluster1 0.402    10 cluster1 (10)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> s499    Phospho_EGF_0min_rep1   13.6  cluster1 0.402    10 cluster1 (10)</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 570 more rows</span></span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tsc</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="SmartPhos_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>A single gene or phospho-site time series data can also be
plotted:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">timerange</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">newSEzero</span><span class="op">$</span><span class="va">timepoint</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotTimeSeries.html">plotTimeSeries</a></span><span class="op">(</span><span class="va">newSEzero</span>, type <span class="op">=</span> <span class="st">"expression"</span>, geneID <span class="op">=</span> <span class="st">"s40"</span>, </span>
<span>               symbol <span class="op">=</span> <span class="st">"RBM47_T519"</span>, condition <span class="op">=</span> <span class="st">"treatment"</span>, </span>
<span>               treatment <span class="op">=</span> <span class="st">"EGF"</span>, timerange <span class="op">=</span> <span class="va">timerange</span><span class="op">)</span></span></code></pre></div>
<p><img src="SmartPhos_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="enrichment-analysis">Enrichment analysis<a class="anchor" aria-label="anchor" href="#enrichment-analysis"></a>
</h2>
<p>Enrichment analysis can be performed on either gene sets
(gene-centric) or post-translational modification signature sets
(site-centric) that are differentially expressed or in a time-series
cluster.</p>
<p>In the latter, each set contains PTM site names with direction of
regulation (up- or down-regulated) instead of gene names. Only
phosphorylation sites will be considered since this pipeline supports
proteomic and phosphoproteomic data. The gene sets and PTM signature
sets are derived from the Molecular Signatures Database (Subramanian,
Tamayo et al.,2005; Liberzon et al., 2011, Liberzon et al., 2015) and
the PTM Signature Database version 2.0.0 (Krug et al., 2019),
respectively. Users are encouraged to consult the PTMsigDB website (<a href="https://proteomics.broadapps.org/ptmsigdb/" class="external-link uri">https://proteomics.broadapps.org/ptmsigdb/</a>) and the
paper of Krug et al. for details on how PTM signature sets were curated
and their annotation.</p>
<p>In gene-centric pathway enrichment, users can perform either
Parametric Analysis of Gene Set Enrichment (PAGE) (Kim &amp; Volsky,
2005) or Gene Set Enrichment Analysis (GSEA) (Subramanian, Tamayo et
al.,2005) with Differential Expression analysis result. With Time series
clustering result, we offer the Fisher’s exact test. For
phospho-signature enrichment on Differential Expression analysis result,
we offer PTM-Signature Enrichment Analysis (PTM-SEA), a method adapted
from GSEA by Krug et al. (2019) to be applicable with the PTM Signature
Database. On time series clustering result, we offer the Fisher’s exact
test, in which each signature set is split into two, one containing
upregulated and the other downregulated phosphosites.</p>
<div class="section level3">
<h3 id="gene-enrichment-analysis-on-differential-expression-analysis-results">Gene Enrichment analysis on differential expression analysis
results<a class="anchor" aria-label="anchor" href="#gene-enrichment-analysis-on-differential-expression-analysis-results"></a>
</h3>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the gene set</span></span>
<span><span class="va">genesetPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"shiny-app/geneset"</span>, package <span class="op">=</span> <span class="st">"SmartPhos"</span><span class="op">)</span></span>
<span><span class="va">inGMT</span> <span class="op">&lt;-</span> <span class="fu">piano</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/piano/man/loadGSC.html" class="external-link">loadGSC</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">genesetPath</span>,<span class="st">"/Cancer_Hallmark.gmt"</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"gmt"</span><span class="op">)</span></span>
<span><span class="co"># Call the function</span></span>
<span><span class="va">resTab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrichDifferential.html">enrichDifferential</a></span><span class="op">(</span>dea <span class="op">=</span> <span class="va">dea</span><span class="op">$</span><span class="va">resDE</span>, type <span class="op">=</span> <span class="st">"Pathway enrichment"</span>, </span>
<span>                             gsaMethod <span class="op">=</span> <span class="st">"PAGE"</span>, geneSet <span class="op">=</span> <span class="va">inGMT</span>, </span>
<span>                             statType <span class="op">=</span> <span class="st">"stat"</span>, nPerm <span class="op">=</span> <span class="fl">200</span>, sigLevel <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>                             ifFDR <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Checking arguments...done!</span></span>
<span><span class="co">## Calculating gene set statistics...done!</span></span>
<span><span class="co">## Calculating gene set significance...done!</span></span>
<span><span class="co">## Adjusting for multiple testing...done!</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTab</span></span></code></pre></div>
<pre><code><span><span class="co">##                       Name Gene Number   Stat      p.up p.up.adj  p.down</span></span>
<span><span class="co">## 1 HALLMARK_APICAL_JUNCTION           1 2.5957 0.0047195  0.10383 0.99528</span></span>
<span><span class="co">## 2 HALLMARK_MITOTIC_SPINDLE           4 2.0511 0.0201290  0.22142 0.97987</span></span>
<span><span class="co">##   p.down.adj Number up Number down</span></span>
<span><span class="co">## 1    0.99528         1           0</span></span>
<span><span class="co">## 2    0.99528         4           0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="gene-enrichment-analysis-on-time-series-results">Gene Enrichment analysis on time-series results<a class="anchor" aria-label="anchor" href="#gene-enrichment-analysis-on-time-series-results"></a>
</h3>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the gene set</span></span>
<span><span class="va">genesetPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"shiny-app/geneset"</span>, package <span class="op">=</span> <span class="st">"SmartPhos"</span><span class="op">)</span></span>
<span><span class="va">inGMT</span> <span class="op">&lt;-</span> <span class="fu">piano</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/piano/man/loadGSC.html" class="external-link">loadGSC</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">genesetPath</span>,<span class="st">"/Chemical_and_Genetic_Perturbations.gmt"</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"gmt"</span><span class="op">)</span></span>
<span><span class="co"># Call the function</span></span>
<span><span class="va">clustEnr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterEnrich.html">clusterEnrich</a></span><span class="op">(</span>clusterTab <span class="op">=</span> <span class="va">tsc</span><span class="op">$</span><span class="va">cluster</span>,</span>
<span>                          se <span class="op">=</span> <span class="va">newSE</span>, inputSet <span class="op">=</span> <span class="va">inGMT</span>,</span>
<span>                          filterP <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                          ifFDR <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">clustEnr</span></span></code></pre></div>
<pre><code><span><span class="co">## $table</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 31 × 9</span></span></span>
<span><span class="co">##    Name          Gene.number Set.size    pval Genes  padj cluster ifSig atLeast1</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lis&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> BROWN_MYELOI…           2        2 0.007<span style="text-decoration: underline;">21</span> <span style="color: #949494;">&lt;chr&gt;</span> 0.794 cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> WANG_CISPLAT…           2        2 0.007<span style="text-decoration: underline;">21</span> <span style="color: #949494;">&lt;chr&gt;</span> 0.794 cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> CREIGHTON_EN…           2        2 0.007<span style="text-decoration: underline;">21</span> <span style="color: #949494;">&lt;chr&gt;</span> 0.794 cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> YOSHIMURA_MA…           2        2 0.007<span style="text-decoration: underline;">21</span> <span style="color: #949494;">&lt;chr&gt;</span> 0.794 cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> JOHNSTONE_PA…           3        6 0.008<span style="text-decoration: underline;">19</span> <span style="color: #949494;">&lt;chr&gt;</span> 1     cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> IVANOVA_HEMA…           3        7 0.013<span style="text-decoration: underline;">8</span>  <span style="color: #949494;">&lt;chr&gt;</span> 0.794 cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ACEVEDO_NORM…           2        2 0.017<span style="text-decoration: underline;">3</span>  <span style="color: #949494;">&lt;chr&gt;</span> 1     cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ACEVEDO_LIVE…           2        2 0.017<span style="text-decoration: underline;">3</span>  <span style="color: #949494;">&lt;chr&gt;</span> 1     cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ACEVEDO_LIVE…           2        2 0.017<span style="text-decoration: underline;">3</span>  <span style="color: #949494;">&lt;chr&gt;</span> 1     cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> FARMER_BREAS…           2        3 0.020<span style="text-decoration: underline;">7</span>  <span style="color: #949494;">&lt;chr&gt;</span> 0.794 cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 21 more rows</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $plot</span></span></code></pre>
<p><img src="SmartPhos_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="phospho-enrichment-analysis-on-time-series-results">Phospho Enrichment analysis on time-series results<a class="anchor" aria-label="anchor" href="#phospho-enrichment-analysis-on-time-series-results"></a>
</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the ptm set</span></span>
<span><span class="va">ptmsetPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"shiny-app/ptmset"</span>, package <span class="op">=</span> <span class="st">"SmartPhos"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">ptmsetPath</span>, <span class="st">"/human_PTM.rda"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Call the function</span></span>
<span><span class="va">clustEnr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterEnrich.html">clusterEnrich</a></span><span class="op">(</span>clusterTab <span class="op">=</span> <span class="va">tsc</span><span class="op">$</span><span class="va">cluster</span>, se <span class="op">=</span> <span class="va">newSE</span>, </span>
<span>                          inputSet <span class="op">=</span> <span class="va">ptmSetDb</span>, ptm <span class="op">=</span> <span class="cn">TRUE</span>, filterP <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                          ifFDR <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">clustEnr</span></span></code></pre></div>
<pre><code><span><span class="co">## $table</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 4 × 9</span></span></span>
<span><span class="co">##   Name           Gene.number Set.size   pval Genes   padj cluster ifSig atLeast1</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lis&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> DISEASE-PSP_P…           1        1 0.040<span style="text-decoration: underline;">7</span> <span style="color: #949494;">&lt;chr&gt;</span> 0.071<span style="text-decoration: underline;">2</span> cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> DISEASE-PSP_p…           1        1 0.040<span style="text-decoration: underline;">7</span> <span style="color: #949494;">&lt;chr&gt;</span> 0.071<span style="text-decoration: underline;">2</span> cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> DISEASE-PSP_n…           1        1 0.040<span style="text-decoration: underline;">7</span> <span style="color: #949494;">&lt;chr&gt;</span> 0.071<span style="text-decoration: underline;">2</span> cluste… TRUE  TRUE    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> DISEASE-PSP_c…           1        1 0.040<span style="text-decoration: underline;">7</span> <span style="color: #949494;">&lt;chr&gt;</span> 0.071<span style="text-decoration: underline;">2</span> cluste… TRUE  TRUE    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $plot</span></span></code></pre>
<p><img src="SmartPhos_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="kinase-activity-inference">Kinase Activity Inference<a class="anchor" aria-label="anchor" href="#kinase-activity-inference"></a>
</h2>
<p>SmartPhos can perform kinase activity inference based on
phosphopeptides that are differentially expressed or in a cluster.
Similar to enrichment analysis, users would first need to perform either
a differential expression analysis or time-series clustering and select
a cluster of interest.</p>
<p>A network of kinase-phosphosite interactions is constructed using the
package <strong>OmnipathR</strong> (Türei et al., 2021). Users can
choose to construct this network with prior knowledge from either Homo
sapiens (taxonomy ID = 9606) or Mus musculus (taxonomy ID = 10090).</p>
<p>By combining prior knowledge about known kinase-phosphosite
interactions and the data, SmartPhos can infer the activity of the
kinases responsible for the phosphopeptides being considered. The
activity is estimated by an activity score computed with the package
<strong>decoupleR</strong> (Badia-I-Mompel et al., 2022) following the
authors’ tutorial on Kinase and Transcription Factor activity
estimation. For time-series clustering, users can also estimate how
likely the kinases are associated with phosphopeptides in the selected
cluster.</p>
<div class="section level3">
<h3 id="kinase-activity-inference-on-differential-expression-analysis-results">Kinase activity inference on differential expression analysis
results<a class="anchor" aria-label="anchor" href="#kinase-activity-inference-on-differential-expression-analysis-results"></a>
</h3>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">netw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getDecouplerNetwork.html">getDecouplerNetwork</a></span><span class="op">(</span><span class="st">"Homo sapiens"</span><span class="op">)</span></span>
<span><span class="va">scoreTab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcKinaseScore.html">calcKinaseScore</a></span><span class="op">(</span><span class="va">dea</span><span class="op">$</span><span class="va">resDE</span>, <span class="va">netw</span>, statType <span class="op">=</span> <span class="st">"stat"</span>, nPerm <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Ventura 13.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Berlin</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] SmartPhos_0.99.0 testthat_3.2.1.1 devtools_2.4.5   usethis_3.0.0   </span></span>
<span><span class="co">## [5] BiocStyle_2.32.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] splines_4.4.1               shinythemes_1.2.0          </span></span>
<span><span class="co">##   [3] later_1.3.2                 norm_1.0-11.1              </span></span>
<span><span class="co">##   [5] bitops_1.0-8                tibble_3.2.1               </span></span>
<span><span class="co">##   [7] preprocessCore_1.66.0       XML_3.99-0.17              </span></span>
<span><span class="co">##   [9] factoextra_1.0.7            lifecycle_1.0.4            </span></span>
<span><span class="co">##  [11] doParallel_1.0.17           rprojroot_2.0.4            </span></span>
<span><span class="co">##  [13] lattice_0.22-6              MASS_7.3-61                </span></span>
<span><span class="co">##  [15] MultiAssayExperiment_1.30.3 magrittr_2.0.3             </span></span>
<span><span class="co">##  [17] limma_3.60.4                plotly_4.10.4              </span></span>
<span><span class="co">##  [19] sass_0.4.9                  rmarkdown_2.28             </span></span>
<span><span class="co">##  [21] jquerylib_0.1.4             yaml_2.3.10                </span></span>
<span><span class="co">##  [23] remotes_2.5.0               shinyBS_0.61.1             </span></span>
<span><span class="co">##  [25] httpuv_1.6.15               doRNG_1.8.6                </span></span>
<span><span class="co">##  [27] sessioninfo_1.2.2           pkgbuild_1.4.4             </span></span>
<span><span class="co">##  [29] cowplot_1.1.3               MsCoreUtils_1.16.1         </span></span>
<span><span class="co">##  [31] RColorBrewer_1.1-3          abind_1.4-5                </span></span>
<span><span class="co">##  [33] pkgload_1.4.0               zlibbioc_1.50.0            </span></span>
<span><span class="co">##  [35] GenomicRanges_1.56.1        purrr_1.0.2                </span></span>
<span><span class="co">##  [37] itertools_0.1-3             AnnotationFilter_1.28.0    </span></span>
<span><span class="co">##  [39] BiocGenerics_0.50.0         sandwich_3.1-0             </span></span>
<span><span class="co">##  [41] proDA_1.18.0                circlize_0.4.16            </span></span>
<span><span class="co">##  [43] GenomeInfoDbData_1.2.12     IRanges_2.38.1             </span></span>
<span><span class="co">##  [45] S4Vectors_0.42.1            ggrepel_0.9.5              </span></span>
<span><span class="co">##  [47] pheatmap_1.0.12             missForest_1.5             </span></span>
<span><span class="co">##  [49] parallelly_1.38.0           MSnbase_2.30.1             </span></span>
<span><span class="co">##  [51] pkgdown_2.1.0               ncdf4_1.23                 </span></span>
<span><span class="co">##  [53] codetools_0.2-20            DelayedArray_0.30.1        </span></span>
<span><span class="co">##  [55] DT_0.33                     tidyselect_1.2.1           </span></span>
<span><span class="co">##  [57] gmm_1.8                     shape_1.4.6.1              </span></span>
<span><span class="co">##  [59] farver_2.1.2                UCSC.utils_1.0.0           </span></span>
<span><span class="co">##  [61] shinyWidgets_0.8.6          matrixStats_1.4.0          </span></span>
<span><span class="co">##  [63] stats4_4.4.1                jsonlite_1.8.8             </span></span>
<span><span class="co">##  [65] GetoptLong_1.0.5            e1071_1.7-14               </span></span>
<span><span class="co">##  [67] ellipsis_0.3.2              iterators_1.0.14           </span></span>
<span><span class="co">##  [69] systemfonts_1.1.0           foreach_1.5.2              </span></span>
<span><span class="co">##  [71] tools_4.4.1                 ragg_1.3.2                 </span></span>
<span><span class="co">##  [73] Rcpp_1.0.13                 glue_1.7.0                 </span></span>
<span><span class="co">##  [75] SparseArray_1.4.8           decoupleR_2.10.0           </span></span>
<span><span class="co">##  [77] xfun_0.47                   piano_2.20.0               </span></span>
<span><span class="co">##  [79] MatrixGenerics_1.16.0       GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">##  [81] dplyr_1.1.4                 shinydashboard_0.7.2       </span></span>
<span><span class="co">##  [83] withr_3.0.1                 BiocManager_1.30.25        </span></span>
<span><span class="co">##  [85] fastmap_1.2.0               fansi_1.0.6                </span></span>
<span><span class="co">##  [87] shinyjs_2.1.0               relations_0.6-13           </span></span>
<span><span class="co">##  [89] caTools_1.18.3              DEP_1.26.0                 </span></span>
<span><span class="co">##  [91] digest_0.6.37               R6_2.5.1                   </span></span>
<span><span class="co">##  [93] mime_0.12                   imputeLCMD_2.1             </span></span>
<span><span class="co">##  [95] textshaping_0.4.0           colorspace_2.1-1           </span></span>
<span><span class="co">##  [97] gtools_3.9.5                utf8_1.2.4                 </span></span>
<span><span class="co">##  [99] tidyr_1.3.1                 generics_0.1.3             </span></span>
<span><span class="co">## [101] data.table_1.16.0           class_7.3-22               </span></span>
<span><span class="co">## [103] PSMatch_1.8.0               httr_1.4.7                 </span></span>
<span><span class="co">## [105] htmlwidgets_1.6.4           S4Arrays_1.4.1             </span></span>
<span><span class="co">## [107] sets_1.0-25                 pkgconfig_2.0.3            </span></span>
<span><span class="co">## [109] gtable_0.3.5                ComplexHeatmap_2.20.0      </span></span>
<span><span class="co">## [111] impute_1.78.0               XVector_0.44.0             </span></span>
<span><span class="co">## [113] brio_1.1.5                  htmltools_0.5.8.1          </span></span>
<span><span class="co">## [115] fgsea_1.30.0                profvis_0.3.8              </span></span>
<span><span class="co">## [117] bookdown_0.40               MALDIquant_1.22.3          </span></span>
<span><span class="co">## [119] ProtGenerics_1.36.0         clue_0.3-65                </span></span>
<span><span class="co">## [121] scales_1.3.0                Biobase_2.64.0             </span></span>
<span><span class="co">## [123] tmvtnorm_1.6                png_0.1-8                  </span></span>
<span><span class="co">## [125] knitr_1.48                  rstudioapi_0.16.0          </span></span>
<span><span class="co">## [127] tzdb_0.4.0                  reshape2_1.4.4             </span></span>
<span><span class="co">## [129] rjson_0.2.22                visNetwork_2.1.2           </span></span>
<span><span class="co">## [131] proxy_0.4-27                cachem_1.1.0               </span></span>
<span><span class="co">## [133] zoo_1.8-12                  GlobalOptions_0.1.2        </span></span>
<span><span class="co">## [135] stringr_1.5.1               KernSmooth_2.23-24         </span></span>
<span><span class="co">## [137] parallel_4.4.1              miniUI_0.1.1.1             </span></span>
<span><span class="co">## [139] vipor_0.4.7                 mzID_1.42.0                </span></span>
<span><span class="co">## [141] vsn_3.72.0                  desc_1.4.3                 </span></span>
<span><span class="co">## [143] pillar_1.9.0                grid_4.4.1                 </span></span>
<span><span class="co">## [145] vctrs_0.6.5                 gplots_3.1.3.1             </span></span>
<span><span class="co">## [147] pcaMethods_1.96.0           slam_0.1-53                </span></span>
<span><span class="co">## [149] randomForest_4.7-1.1        urlchecker_1.0.1           </span></span>
<span><span class="co">## [151] promises_1.3.0              xtable_1.8-4               </span></span>
<span><span class="co">## [153] cluster_2.1.6               beeswarm_0.4.0             </span></span>
<span><span class="co">## [155] evaluate_0.24.0             readr_2.1.5                </span></span>
<span><span class="co">## [157] mvtnorm_1.3-1               cli_3.6.3                  </span></span>
<span><span class="co">## [159] compiler_4.4.1              rngtools_1.5.2             </span></span>
<span><span class="co">## [161] rlang_1.1.4                 crayon_1.5.3               </span></span>
<span><span class="co">## [163] labeling_0.4.3              marray_1.82.0              </span></span>
<span><span class="co">## [165] QFeatures_1.14.2            affy_1.82.0                </span></span>
<span><span class="co">## [167] plyr_1.8.9                  fs_1.6.4                   </span></span>
<span><span class="co">## [169] ggbeeswarm_0.7.2            stringi_1.8.4              </span></span>
<span><span class="co">## [171] viridisLite_0.4.2           BiocParallel_1.38.0        </span></span>
<span><span class="co">## [173] assertthat_0.2.1            munsell_0.5.1              </span></span>
<span><span class="co">## [175] lazyeval_0.2.2              Matrix_1.7-0               </span></span>
<span><span class="co">## [177] hms_1.1.3                   ggplot2_3.5.1              </span></span>
<span><span class="co">## [179] statmod_1.5.0               shiny_1.9.1                </span></span>
<span><span class="co">## [181] highr_0.11                  SummarizedExperiment_1.34.0</span></span>
<span><span class="co">## [183] mzR_2.38.0                  igraph_2.0.3               </span></span>
<span><span class="co">## [185] memoise_2.0.1               affyio_1.74.0              </span></span>
<span><span class="co">## [187] bslib_0.8.0                 fastmatch_1.1-4</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Shubham Agrawal, Junyan Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
