<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Perform Cluster Enrichment Analysis — clusterEnrich • SmartPhos</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform Cluster Enrichment Analysis — clusterEnrich"><meta name="description" content="clusterEnrich performs enrichment analysis on gene clusters, using Fisher's Exact Test to determine the significance of enrichment for each cluster."><meta property="og:description" content="clusterEnrich performs enrichment analysis on gene clusters, using Fisher's Exact Test to determine the significance of enrichment for each cluster."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SmartPhos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SmartPhos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/SmartPhos_Shiny.html">SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Perform Cluster Enrichment Analysis</h1>

      <div class="d-none name"><code>clusterEnrich.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>clusterEnrich</code> performs enrichment analysis on gene clusters, using Fisher's Exact Test to determine the significance of enrichment for each cluster.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">clusterEnrich</span><span class="op">(</span></span>
<span>  <span class="va">clusterTab</span>,</span>
<span>  <span class="va">se</span>,</span>
<span>  <span class="va">inputSet</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ptm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  adj <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  filterP <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  ifFDR <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-clustertab">clusterTab<a class="anchor" aria-label="anchor" href="#arg-clustertab"></a></dt>
<dd><p>A <code>data frame</code> containing cluster information, where each row corresponds to a gene and its assigned cluster.</p></dd>


<dt id="arg-se">se<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>A <code>SummarizedExperiment</code> object containing gene expression data and metadata.</p></dd>


<dt id="arg-inputset">inputSet<a class="anchor" aria-label="anchor" href="#arg-inputset"></a></dt>
<dd><p>A <code>list</code> or <code>data frame</code> of gene sets to be used for enrichment analysis.</p></dd>


<dt id="arg-reference">reference<a class="anchor" aria-label="anchor" href="#arg-reference"></a></dt>
<dd><p>A <code>character</code> vector of reference genes. If NULL, it will be extracted from se object. Default is <code>NULL</code>.</p></dd>


<dt id="arg-ptm">ptm<a class="anchor" aria-label="anchor" href="#arg-ptm"></a></dt>
<dd><p><code>Logical</code>. If <code>TRUE</code>, the function will perform enrichment analysis on post-translational modification (PTM) gene sets. Default is <code>FALSE</code>.</p></dd>


<dt id="arg-adj">adj<a class="anchor" aria-label="anchor" href="#arg-adj"></a></dt>
<dd><p><code>Character</code>. The method for adjusting p-values. Default is "BH".</p></dd>


<dt id="arg-filterp">filterP<a class="anchor" aria-label="anchor" href="#arg-filterp"></a></dt>
<dd><p><code>Numeric</code>. The p-value threshold for filtering significant results. Default is 0.05.</p></dd>


<dt id="arg-iffdr">ifFDR<a class="anchor" aria-label="anchor" href="#arg-iffdr"></a></dt>
<dd><p><code>Logical</code>. If <code>TRUE</code>, the function will use FDR-adjusted p-values for significance filtering. Default is <code>FALSE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>list</code> containing two elements:</p><ul><li><p>`table`: A <code>data frame</code> with enrichment results for each cluster and pathway.</p></li>
<li><p>`plot`: A <code>ggplot2</code> object showing the significance of enrichment for each pathway across clusters.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function first retrieves or computes the reference set of genes or PTM sites. It then performs enrichment analysis for each cluster using the <code>runFisher</code> function.
The results are filtered based on the p-value threshold and adjusted for multiple testing if <code>ifFDR</code> is <code>TRUE</code>. The function generates a dot plot where the size and color of the points represent the significance of enrichment.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load multiAssayExperiment object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dia_example"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Get SummarizedExperiment object</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">dia_example</span><span class="op">[[</span><span class="st">"Phosphoproteome"</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">dia_example</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">seProcess</span> <span class="op">&lt;-</span> <span class="fu"><a href="preprocessPhos.html">preprocessPhos</a></span><span class="op">(</span>seData <span class="op">=</span> <span class="va">se</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span>, impute <span class="op">=</span> <span class="st">"QRILC"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing along margin 2 (samples/columns).</span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="addZeroTime.html">addZeroTime</a></span><span class="op">(</span><span class="va">seProcess</span>, condition <span class="op">=</span> <span class="st">"treatment"</span>, treat <span class="op">=</span> <span class="st">"EGF"</span>, zeroTreat <span class="op">=</span> <span class="st">"1stCrtl"</span>, timeRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"20min"</span>,<span class="st">"40min"</span>, <span class="st">"6h"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Get the numeric matrix</span></span></span>
<span class="r-in"><span><span class="va">exprMat</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Call the clustering function</span></span></span>
<span class="r-in"><span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="clusterTS.html">clusterTS</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">exprMat</span>, k <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genesetPath</span> <span class="op">&lt;-</span> <span class="va">appDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"shiny-app/geneset"</span>, package <span class="op">=</span> <span class="st">"SmartPhos"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">inGMT</span> <span class="op">&lt;-</span> <span class="fu">piano</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/piano/man/loadGSC.html" class="external-link">loadGSC</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">genesetPath</span>,<span class="st">"/Cancer_Hallmark.gmt"</span><span class="op">)</span>,type<span class="op">=</span><span class="st">"gmt"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Call the function</span></span></span>
<span class="r-in"><span><span class="fu">clusterEnrich</span><span class="op">(</span><span class="va">clust</span><span class="op">$</span><span class="va">cluster</span>, <span class="va">seProcess</span>, <span class="va">inGMT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $table</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 0 × 9</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 9 variables: Name &lt;chr&gt;, Gene.number &lt;int&gt;, Set.size &lt;int&gt;, pval &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   Genes &lt;list&gt;, padj &lt;dbl&gt;, cluster &lt;chr&gt;, ifSig &lt;lgl&gt;, atLeast1 &lt;lgl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $plot</span>
<span class="r-plt img"><img src="clusterEnrich-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Shubham Agrawal, Junyan Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

