<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data • SmartPhos</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SmartPhos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.28</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/SmartPhos.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/SmartPhos_Shiny.html">SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Bioconductor/SmartPhos/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>SmartPhos Explorer: a one-stop data analysis platform for proteomic and phosphoproteomic data</h1>
                        <h4 data-toc-skip class="author">Shubham
Agrawal, Cong Quan Ta, Junyan Lu</h4>
            
            <h4 data-toc-skip class="date">2025-03-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/SmartPhos/blob/main/vignettes/SmartPhos_Shiny.Rmd" class="external-link"><code>vignettes/SmartPhos_Shiny.Rmd</code></a></small>
      <div class="d-none name"><code>SmartPhos_Shiny.Rmd</code></div>
    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>SmartPhos Explorer is an R Shiny app for streamlined analysis of
<em>proteomics</em> and <em>phosphoproteomics</em>, featuring data
pre-processing, exploratory data analysis, hypothesis testing,
time-series analysis and functional interpretation. There are two ways
to run the SmartPhos Explorer shiny app:</p>
<ul>
<li>Calling the function after loading the SmartPhos package.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://lu-group-ukhd.github.io/SmartPhos/">"SmartPhos"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/runSmartPhos.html">runSmartPhos</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Users can make an independent directory for the SmartPhos Explorer
app and open any of the script (server.R, ui.R) in RStudio and click on
<em>Run App</em> button.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://lu-group-ukhd.github.io/SmartPhos/">"SmartPhos"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/makeSmartPhosDirectory.html">makeSmartPhosDirectory</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"shinyApp"</span><span class="op">)</span></span></code></pre></div>
<p>The front page of the Shiny app looks as below:</p>
<p><img src="images/firstLook.png" width="1138"></p>
<p>The current tabs are:</p>
<ul>
<li>Preprocessing</li>
<li>PCA</li>
<li>Heatmap</li>
<li>Differential expression</li>
<li>Time-series clustering</li>
<li>Enrichment analysis</li>
<li>Kinase activity inference</li>
<li>Log Info</li>
<li>Update Notes</li>
</ul>
<p>In this vignette, we will walk you through the process of preparing
your data and performing downstream analyses using different tabs in the
Shiny app.</p>
</div>
<div class="section level2">
<h2 id="preprocessing-tab">Preprocessing tab<a class="anchor" aria-label="anchor" href="#preprocessing-tab"></a>
</h2>
<p>This tab contains the functions for preprocessing before diving into
data analysis and interpretation. In this tab, users can start with
uploading a MultiAssayExperiment object, generated by the SmartPhos R
package or a zip file containing quantification tables from Spectronaut
or MaxQuant. Output from DIA-NN is planned for the future update. After
uploading the data, options for choosing the transformation and
normalization methods, performing batch effect correction, and
displaying the missing values are available. There is also an option for
saving the uploaded data as R object (.rds file) for reusing in the
future to save processing time.</p>
<div class="section level3">
<h3 id="upload-options">Upload options<a class="anchor" aria-label="anchor" href="#upload-options"></a>
</h3>
<p>Uploading is one of the most important steps in the smooth
functioning of the app.</p>
<p><img src="images/preprocess1.png" width="400px"></p>
<p>As shown in the figure, users have the option to decide between
uploading a <strong>zip file</strong> and a <strong>MultiAssayExperiment
object</strong>. The zip file contains the output files generated by
<em>Spectronaut</em> or <em>MaxQuant</em> software. For users who are
not experienced with R, this is the preferred option.</p>
<p>When using the zip file option, there are certain requirements the
users need to follow for the error-free conversion of the data from
these files into a multiAssayExperiment R object. One of the most
important requirements for the zip file is to contain a properly
formatted <strong>fileTable.txt</strong> file. The details of this file
are explained in the next subsection.</p>
<div class="section level4">
<h4 id="filetable-txt-file">fileTable.txt file<a class="anchor" aria-label="anchor" href="#filetable-txt-file"></a>
</h4>
<p>This file should be a tab-separated <strong>tsv</strong> file that
contains the information including name of the quantification table,
sample ID and experimental details, such as whether a certain sample was
enriched for phospho-peptides, whether the quantification was on protein
groups or phosphorylation sites and so on.<br>
Optional annotations, such experimental conditions, time points and etc.
can also be included in this table and some of them are required for
specific downstream analysis tasks. An error message will be displayed
if this file is missing or the format of file is incorrect. The
mandatory columns in this table are as follows:</p>
<ul>
<li><p><strong>fileName</strong>: The names of quantification tables of
protein groups or phosphorylation sites. They should be the standard
output files from Spectronaut or MaxQuant. Make sure that it contains
only the file names with the file extension (For example:
“20240826_Protein_Report.xls”, “20240826_PTMReport.xls”)) and not the
relative or absolute path.</p></li>
<li><p><em>Important notes on the format of the quantification table
from Spectronaut:</em> The quantification table from Spectronaut should
be in the “wide” format, with different samples in different columns.
The table should be in plain text format, not Excel tables. The file
extension does not matter. Spectronaut sometimes output .xls files but
they are actually plain text files. The acceptable deliminators for the
tables are: tab ( (preferred), comma (,), colon (:) and semicolon (;).
For full proteome measurement, “PG.ProteinGroups”, “PG.Genes” and
“[sample name]PG.Quantity” columns must be present. For phosphoproteome
measurement, “[sample name]PTM.SiteProbability”, “[sample
name]PTM.Quantity”, “PTM.CollapseKey”, “PG.UniProtIds”,“PG.Genes”,
“PTM.SiteLocation”, “PTM.SiteAA”, and “PTM.FlankingRegion” columns must
be present.</p></li>
<li><p><strong>id</strong>: Contains unique identifiers for sample
(sample IDs). The sample IDs should correspond to the prefix of actual
column names in the quantification table. For example: the ID value
<strong>FullProteome_1stCrtl_0min_rep2</strong> corresponds to one of
the column names
oecf4_220602_BH_ET_<strong>FullProteome_1stCrtl_0min_rep2</strong>.raw.PG.Quantity
in the quantification table. SmartPhos will automatically recognize the
columns in the quantification table that match a certain sample ID, as
long as this sample ID is present in the columns that contain
“PG.Quantity”.</p></li>
<li><p><strong>searchType</strong>: Contains information on the type of
search performed on the samples by Spectronaut or MaxQuant. There are
two possibilities: <strong>proteome</strong> (quantification of protein
abundance) and <strong>phosphoproteome</strong> (quantification of the
abundance of phosphorylation sites). The two searchTypes are represented
as two assays in the multiAssayExperiment object. Note that the app will
still work if only one of the searchType is present but some functions
may not be available. For example, normalisation of phosphorylation
sites by protein abundance will be not available of only phosphoproteome
measurement is available.</p></li>
<li><p><strong>sampleType</strong>: Contains information of whether
enrichment for phosphorylated peptides is performed for a certain
sample. The sampleType can be <strong>Phospho</strong> (or
<strong>PP</strong>) for the phospho-enriched samples and
<strong>FullProteome</strong> (or <strong>FP</strong>) for the
non-enriched sample. Not that the app will still work if only one of the
sample type is present but some functions may not be available. For
example, the novel normalisation correction for phosphoproteomic data
will not be available if full proteome without enrichment is
absent.</p></li>
</ul>
<p>The above mentioned columns are mandatory ones. Other than that,
fileTable.txt can contain optional columns such as:</p>
<ul>
<li><p><strong>sampleName</strong> column is important for performing
normalization correction or for normalizing Phosphoproteome samples by
Full Proteome samples. This should be the sample id without FP_/PP_ or
Phospho_/FullProteome_ suffix or prefix. For example, for the sample ID,
<strong>FullProteome_1stCrtl_0min_rep1</strong>, the sampleName should
be, “<strong>1stCrtl_0min_rep1</strong>”. Without this column, it would
not be possible to perform normalization correction and normalizing
phosphoproteome samples by full proteome samples.</p></li>
<li><p><strong>subjectID</strong> with information like patientID, cell
lines, etc. This column will be used if there are multiple samples from
the same subject and users wish to include the subject ID as covariates
in hypothesis testing (e.g. paired t-test) or time-series
analysis.</p></li>
<li><p><strong>timepoint</strong> with values like 0min, 20min, 6h, 24h,
etc. This column is important for the time-series clustering tab and
other tabs which make use of the results from the time-series clustering
tab.</p></li>
<li><p><strong>replicate</strong> with values rep1, rep2, replicate1 and
so on.</p></li>
<li><p><strong>treatment</strong> containing information about different
treatments or experimental conditions.</p></li>
<li><p>Users can include any metadata for visualization or hypothesis
testing in the fileTable.txt.</p></li>
</ul>
<p>Here is an example of the fileTable.txt file:</p>
<p><img src="images/preprocess2.png" width="600px"></p>
<p><strong>NOTE</strong>: It’s better to avoid any special characters
(‘(’, ‘)’, ‘#’, ‘%’, ‘@’, etc) in the column names in the quantification
table output by Spectronaut/MaxQuant and same for the ID column of the
fileTable.txt file. In addition, sample IDs <strong>should not</strong>
start with numbers. Those can cause error in parsing the data.</p>
</div>
<div class="section level4">
<h4 id="column-annotations">Column annotations<a class="anchor" aria-label="anchor" href="#column-annotations"></a>
</h4>
<p>Users have option to select the column annotations, which are
provided by users in the fileTable.txt, when using the zip file method.
If the upload of zip file is successful, another widget will appear
which will allow the user to select from the different available
annotation options. Multiple selections are possible simply by clicking
on the list item.</p>
<p><img src="images/preprocess3.png" width="350px"></p>
</div>
<div class="section level4">
<h4 id="multiassayexperiment-object">MultiAssayExperiment object<a class="anchor" aria-label="anchor" href="#multiassayexperiment-object"></a>
</h4>
<p>For users with experience in R, they can parse data by using our
SmartPhos package, which will create an multiAssayExperiment object.
Users can save this object as an <strong>rds</strong> file and choose
the “MultiAssayExperiment object” option for uploading the data.</p>
<p>Users can also first upload the zip file to SmartPhos Explorer and
download the processed data as an <strong>rds</strong> file (see section
2.1.5) . This file can be uploaded using the “MultiAssayExperiment
option** option in the future analysis, which will save some time as
processing the quantification table can be slow if the file size is
large.</p>
</div>
<div class="section level4">
<h4 id="error-checks">Error checks<a class="anchor" aria-label="anchor" href="#error-checks"></a>
</h4>
<p>If the processing of zip files or multiAssayExperiment object is
successful, then the upload tab will show the total number of samples
and features.</p>
<p><img src="images/preprocess4.png" width="350px"></p>
<p>The user will have option to choose the assay and sample type as
shown in the figure above. In case the upload is not successful, the app
will display an error message. These error messages are:</p>
<p><img src="images/error1.png" width="350px"><img src="images/error2.png" width="350px"><img src="images/error3.png" width="350px"></p>
</div>
<div class="section level4">
<h4 id="save-and-re-use-the-parsed-dataset-">Save and re-use the parsed dataset.<a class="anchor" aria-label="anchor" href="#save-and-re-use-the-parsed-dataset-"></a>
</h4>
<p>Users have the options to download the parsed data as a .rds file for
future use. There are also options of saving the parsed data in the
local version of Shiny app as well as to load and delete previously
saved data objects.</p>
</div>
</div>
<div class="section level3">
<h3 id="preprocessing-options">Preprocessing options<a class="anchor" aria-label="anchor" href="#preprocessing-options"></a>
</h3>
<p>SmartPhos package performs some pre-filtering and preprocessing based
on various threshold values during the generation of the
multiAssayExperiment object. On top of that, this panel has different
options for preprocessing of the selected assay. The options provided by
the shiny app are as follows:</p>
<ul>
<li>
<strong>Normalization correction</strong>: This option is present
only if the phophoproteomics data is present. It allows for the
correction of normalization artefacts introduced by the
Spectronaut.</li>
<li>
<strong>Normalize</strong> phospho intensity by the corresponding
protein expression.</li>
<li>
<strong>Transformation</strong>: The transformation methods
available are: log2 and vst (variance stabilizing transformation).</li>
<li>
<strong>Normalization</strong>: Normalization strategy depends on
the selected transformation method. Therefore, the user has option of
<em>Yes</em> or <em>No</em>. If the user selects <em>Yes</em>, then for
log2 and no transformation, <em>median scaling</em> is applied and for
vst, <em>vsn</em> is applied.</li>
<li>
<strong>Missing values</strong>: Users can select what percentage of
missing values are allowed. Proteins with missing values above the
selected threshold will be removed from all analyses.</li>
<li>
<strong>Imputation</strong>: Currently four imputation methods are
available:
<ul>
<li>QRILC (Quantile Regression Imputation of Left-Censored data)</li>
<li>MinDet (Deterministic minimal value approach)</li>
<li>BPCA (Bayesian PCA)</li>
<li>MLE (Maximum Likelihood Estimation)</li>
<li>Random forest</li>
</ul>
</li>
<li>
<strong>Batch effects removal</strong>: This option provides the
abilty to correct for batch effects. It uses
<em>removeBatchEffect()</em> from the <em>limma</em> package. The users
can select the maximum of two columns. The columns used for batch
effects removal should be present in the fileTable.txt file.</li>
</ul>
<p>After selecting the options, press the <strong>Process</strong>
button to perform the preprocessing.</p>
<p><img src="images/preprocess.png" width="400px"></p>
</div>
<div class="section level3">
<h3 id="other-options">Other options<a class="anchor" aria-label="anchor" href="#other-options"></a>
</h3>
<p>The preprocessing tab has some more options:</p>
<ul>
<li>Plot the <strong>completeness of the assay</strong>. It will plot a
bar plot displaying percentage of completeness for each sample.</li>
<li>Launching MatrixQCvis. Another shiny app for exploratory
analysis.</li>
<li>Saving, loading and removing results. The multiAssayExperiment
object can be saved, loaded and removed within the shiny app.</li>
<li>
<strong>Removing outlier</strong>. Simply writing the names of the
samples (delimited by comma) in the text box will remove those samples.
It also supports regex option. For example: if “100min” is typed in the
outlier box, then all the samples with 100min timepoint will be
removed.</li>
<li>Coloring the box plot based on different annotations is also
possible.</li>
<li>
<strong>Subsetting</strong> of the dataset is possible from the
displayed table. The selected annotations in the filter options of
different columns will be used for subsetting the data. The following
tabs of the shinyApp will use that subsetted data for further
exploration and analysis. Subsetting can be undone by simply removing
the annotations from filter boxes.</li>
<li>After preprocessing, the assay is saved as a
<strong>SummarizedExperiment</strong> object and user can download it as
a rds file.</li>
</ul>
<p><img src="images/otherOptions1.png" width="300px"></p>
<p><img src="images/otherOptions2.png" width="600px"></p>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>The different outputs available to the users are:</p>
<ul>
<li>A box plot of intensity values for each samples.</li>
<li>A diagnostic plot of phopho to fullProteome ratio. This plot can be
used to decide if the normalization correction should be performed or
not. It can also shows the outliers.</li>
<li>A data table showing the information about each samples.</li>
<li>A bar plot which shows the percentage of completness of each
samples.</li>
</ul>
<p><img src="images/boxPlot.png" width="973"><img src="images/logRatio.png" width="973"><img src="images/table.png" width="985"><img src="images/completeness.png" width="981"></p>
</div>
</div>
<div class="section level2">
<h2 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h2>
<p>This tab performs principal component analysis (PCA) on the imputed
assay from the <em>Preprocessing</em> tab and then plot the principal
components.</p>
<p><img src="images/pca1.png" width="300px"></p>
<p>Simply click on <strong>Run PCA</strong> button to perform PCA. The
different options available to the users after plotting are:</p>
<p><img src="images/pca2.png" width="300px"></p>
<ul>
<li>Selecting the principal components for x- and y-axis.</li>
<li>Two more dimensions can be added to the data points using
<strong>color</strong> and <strong>shape</strong> options.</li>
<li>The plotting is done using <strong>Plotly</strong> package, so users
have many functionalities. For example, hovering over the plotted data
points will display more information, downloading plot as png file,
etc.</li>
<li>PCA values can be downloaded as a tsv file.</li>
<li>PCA plot can also be downloaded as a pdf file.</li>
</ul>
<p><img src="images/pca3.png" width="490px"><img src="images/pca4.png" width="600px"><img src="images/pca5.png" width="600px"></p>
<p>NOTE: If imputation is selected <em>none</em> in the preprocessing
tab, then performing PCA is not possible. An error message will pop-up
on screen.</p>
</div>
<div class="section level2">
<h2 id="heatmap">Heatmap<a class="anchor" aria-label="anchor" href="#heatmap"></a>
</h2>
<p>This tabs allow the user to plot the heatmap of the imputed assay
from the preprocessing tab. Users can select from the three choices
available:</p>
<p><img src="images/heatmap.png" width="500px"></p>
<ul>
<li>
<strong>Top variant</strong>: This allows the users to plot the
genes with highest variance. Users can decide the number of top variants
genes to plot. This option performs clustering automatically. The user
also has the option to divide the columns and rows of the heatmap into
specific number of clusters.</li>
<li>
<strong>Differentially expressed</strong>: Allows to plot the
heatmaps for differentially expressed genes. The differential expression
analysis is available in <em>Differential expression</em> tab and can be
performed using ProDA or Limma.</li>
<li>
<strong>Selected time series cluster</strong>: After performing the
time-series clustering in another tab, users have option to plot the
heatmap of the selected cluster.</li>
</ul>
<p>For all the above-mentioned options, users can add additional column
annotations to the plotted heatmap. Adding more column annotations to
the heatmap can be helpful to understand the heatmap better and see the
patterns in the data more clearly. User can also download the heatmap as
a PDF.</p>
<p><img src="images/heatmapPDF.png" width="300px"></p>
<p><img src="images/heatmap1.png" width="900"><img src="images/heatmap2.png" width="893"></p>
<p>The tab also has error checks. If the user tries to plot the heatmap
of differentially expressed genes or the genes from the selected time
series cluster prior to performing the aforementioned analysis, an error
message will be displayed.</p>
<p><img src="images/errorHeatmap.png" width="400px"></p>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>This tab performs differential expression analysis on the transformed
and normalized assay from the first tab. The goal of performing
differential expression analysis is to quantify the expression levels of
genes between different experimental conditions using statistical
tests.</p>
<p>If the users want paired t-test on the patient IDs, cell lines, etc,
then the users must have <strong>subjectID</strong> as one of the column
in the <em>fileTable.txt</em> file with the relevant information. The
subjectID column should also be selected in the additional column
annotations before the generation of multiAssayExperiment object.</p>
<p>The two methods available for performing differential expression
analysis are:</p>
<ul>
<li>
<strong>limma</strong>: uses linear models.</li>
<li>
<strong>ProDA</strong>: uses probabilistic dropout model.</li>
</ul>
<p>To use this tab, first the user has to select the metadata column for
which they want to perform the analysis. Users need to select samples
for the reference and target group, between which the differential
expression analysis will be performed. This can be done by either
selecting treatments or selecting sample IDs. In the former, time points
should also be specified if the column <strong>timepoint</strong> is
present in the <em>fileTable.txt</em> file. Multiple treatments and time
points can be selected for each group. As the selection is updated for
each group, the user interface informs users of the number of samples in
each group as well as a warning if a sample is present in both group.
The latter option allows users to directly choose which sample to
include in each group, thereby allowing more flexibility in sample
selection. The two figures below show examples of the user interface
when using either option. Notice that a warning was given in the latter
since one sample was present in both groups.</p>
<p><img src="images/diff_option1.png" width="350px"></p>
<p><img src="images/diff_option2.png" width="350px"></p>
<p>The output of differential expression analysis is a table which
contains the differentially expressed genes arranged by the lowest
p-values. The other information present in the table are: Uniprot ID,
log2 fold change, t-statistic, adjusted p-value and some information if
the data is phosphoproteomic data. A volcano plot is also plotted,
highlighing the points with positive and negative log fold-change and
above certain p value. The histogram of p-value is also plotted.</p>
<p><img src="images/diff.png" width="1008"></p>
<p>Users have options to filter the differentially expressed genes table
based on p-value and log fold change value. There is also an option of
using adjusted p-value instead.</p>
<p><img src="images/diff_filter.png" width="300px"></p>
<p>If the user click on the row of the differential expression table,
the corresponding point in the volcano plot is highlighted as star.
Moreover, a box plot is plot to show the change in normalized
intensities for the selected conditions. Also, if the user click on any
coloured point in the volcano plot, the corresponding row in the
differential expression table is highlighted. The user also has the
option to download the differential expression table as a TSV file.</p>
<p><img src="images/diff_click.png" width="1008"></p>
<p>In case the differential expression analysis is not possible for the
selected settings, an error message will appear.</p>
<p><img src="images/errorDEA.png" width="400px"></p>
</div>
<div class="section level2">
<h2 id="time-series-clustering">Time series clustering<a class="anchor" aria-label="anchor" href="#time-series-clustering"></a>
</h2>
<p>This tab performs <strong>fuzzy c-means clustering</strong> to group
proteins/phosphopeptides based on how their level changes over time. The
algorithm considers the time-resolved trend, but not the expression
levels. Thus, members of the same cluster would have a similar trend
over time (e.g., all increasing or all decreasing), though their
expression level can be different.</p>
<p>To use this tab, users must specify the <strong>timepoint</strong>
column in the <strong>fileTable.txt</strong> file when preparing the
data. The time points are either unit-less numbers (e.g., 1, 2, 3) or
are in hour and/or minute, which must be typed as “h” and “min”,
respectively (e.g., 1min, 2min, 3h). Please notice that mixing the two
said options (e.g., 1, 2min, 3) or using other unit for time (e.g.,
1hour, 2minute, 3day) will likely lead to wrong results.</p>
<p>The options for using this tab are as follow:</p>
<p><img src="images/timeSeriesOptions.png" width="350px"></p>
<p>Users first need to choose the metadata column of interest for the
analysis. Based on the selected metadata, user will be provided with the
different conditions to choose from. Additionally, if the clustering is
performed on either “logFC” or “two-condition expression”, a reference
condition should also be chosen. The options to perform clustering on
determine what it means for genes to be in the same cluster:</p>
<ul>
<li>
<strong>expression</strong>: Genes in the same cluster have the
similar trend in their <strong>expression level</strong>.</li>
<li>
<strong>logFC</strong>: Genes in the same cluster have the similar
trend in their fold-change, (<strong>logFC</strong>). The fold-change is
the difference in expression level between the two selected
treatments.</li>
<li>
<strong>two-condition expression</strong>: Genes in the same cluster
have the similar trend in their <strong>expression level</strong> and
between the two treatments.</li>
</ul>
<p>Users can choose to include or exclude certain time points by
checking the boxes. The time points are updated when selecting the
treatments.</p>
<div class="section level3">
<h3 id="other-parameters-for-the-analysis">Other parameters for the analysis<a class="anchor" aria-label="anchor" href="#other-parameters-for-the-analysis"></a>
</h3>
<ul>
<li><strong>Use top % variant genes along time</strong></li>
<li>
<strong>Filter genes based on spline fit test</strong>: If selected,
a spline fitting would be used to filter out genes whose changes are
inconsistent. <strong>subjectID</strong> is used to pair samples if
provided, otherwise the replicates are considered independent.</li>
<li>
<strong>Number of clusters</strong>: We advise users to try the
analysis with multiple values to select the best number of
clusters.</li>
<li>
<strong>Cut-off for cluster membership probability</strong>: Genes
are removed from a cluster if their probability to be in that cluster is
below this cut-off value.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output-1">Output<a class="anchor" aria-label="anchor" href="#output-1"></a>
</h3>
<p>The plot shows the time-course trend of genes in the clusters. If
either <strong>expression</strong> or <strong>logFC</strong> was chosen,
the color would indicate the genes’ membership probability:
<img src="images/timeSeriesClust1.png" width="716"></p>
<p>If <strong>two-condition expression</strong> was chosen, the color is
used to distinguish the treatments:
<img src="images/timeSeriesClust2.png" width="716"></p>
<p>Accompanying the plot is a table show the details of the genes in
each cluster. User can select the cluster to view with the drop-down
menu on the upper left side of the table. The table can be downloaded as
a tsv file. The selected cluster will be used as input for enrichment
analysis and kinase activity inference.
<img src="images/time_series_result2.png" width="976"></p>
<p>Clicking on a gene will show a plot of its time-course expression
level and a line showing the average expression level:
<img src="images/time_series_result3.png" width="917"></p>
<p>In case zero timepoint is not available for a selected treatment,
user has the option to add zero timepoint to the treatment from one of
the controls (or conditions with zero timepoint). This will be reflected
in the plots showing the clustering results and also when a particular
gene is selected for seeing the normalized expression level.</p>
<p><img src="images/clustering0.png" width="300px"></p>
<p><img src="images/clustering1.png" width="726"><img src="images/clustering2.png" width="514"></p>
<p>If no cluster was found, for instance due to cut-off values being too
high, users will be informed with a pop-up window:
<img src="images/timeSeriesErr1.png" width="716"></p>
</div>
</div>
<div class="section level2">
<h2 id="enrichment-analysis">Enrichment analysis<a class="anchor" aria-label="anchor" href="#enrichment-analysis"></a>
</h2>
<p>This tab performs enrichment analysis on genes or phosphosites that
are differentially expressed or in a time series cluster. To use this
tab, users would first need to perform either a differential expression
analysis or time-series clustering. Users have three options to choose
the source of gene list:</p>
<ul>
<li>Differential expression: the gene list of the differential analysis
result.</li>
<li>Select time-series cluster: the selected cluster after performing
the time-series clustering.</li>
<li>All time-series cluster: on all the clusters resulted after
performing the time-series clustering.</li>
</ul>
<p>We offer the possibilities to perform enrichment analysis on either
gene sets (gene-centric) or post-translational modification signature
sets (site-centric). In the latter, each set contains PTM site names
with direction of regulation (up- or down-regulated) instead of gene
names. Only phosphorylation sites will be considered since this pipeline
supports proteomic and phosphoproteomic data. The gene sets and PTM
signature sets are derived from the Molecular Signatures Database
(Subramanian, Tamayo et al.,2005; Liberzon et al., 2011, Liberzon et
al., 2015) and the PTM Signature Database version 2.0.0 (Krug et al.,
2019), respectively. Users are encouraged to consult the PTMsigDB
website (<a href="https://proteomics.broadapps.org/ptmsigdb/" class="external-link uri">https://proteomics.broadapps.org/ptmsigdb/</a>) and the
paper of Krug et al. for details on how PTM signature sets were curated
and their annotation.</p>
<p>In gene-centric pathway enrichment, users can perform either
Parametric Analysis of Gene Set Enrichment (PAGE) (Kim &amp; Volsky,
2005) or Gene Set Enrichment Analysis (GSEA) (Subramanian, Tamayo et
al.,2005) with Differential Expression analysis result. With Time series
clustering result, we offer the Fisher’s exact test. For
phospho-signature enrichment on Differential Expression analysis result,
we offer PTM-Signature Enrichment Analysis (PTM-SEA), a method adapted
from GSEA by Krug et al. (2019) to be applicable with the PTM Signature
Database. On time series clustering result, we offer the Fisher’s exact
test, in which each signature set is split into two, one containing
upregulated and the other downregulated phosphosites.</p>
<p>Options for performing enrichment analysis are shown below:
<img src="images/enrich_option.png" width="350px"></p>
<ul>
<li>
<strong>Select analysis method</strong>: Whether to perform
gene-centric pathway enrichment or site-centric phospho-signature
enrichment. If the Proteome assay is selected in the preprocessing tab,
only the former is available.</li>
<li>
<strong>Source of gene list</strong>: Whether to use result from
differential expression analysis, selected time-series cluster or all
time-series cluster</li>
<li>
<strong>Select enrichment method</strong>: Which method should be
used for the analysis. In pathway enrichment, the method is either
“PAGE” (Parametric Analysis of Gene Set Enrichment) or “GSEA” (Gene Set
Enrichment Analysis) for differential expression result and only
Fisher’s exact test for time-series clustering. If “GSEA” is chosen, the
number of permutations to generate the null distribution should also be
specified (default 100). In phospho-signature enrichment, the method is
PTM-SEA (not shown on UI) for differential expression and Fisher’s exact
test for time series clustering.</li>
<li>
<strong>Statistic used for ranking</strong>: Whether to use
t-statistics or log2FC for the analysis, only applies to differential
expression result.</li>
<li>
<strong>Select geneset database</strong>: User here has option
either to select from available gene set/ PTM set database listed in the
box or upload another gene set/ PTM set.</li>
<li>P-value cut-off and option to use FDR.</li>
</ul>
<div class="section level3">
<h3 id="output-2">Output<a class="anchor" aria-label="anchor" href="#output-2"></a>
</h3>
<p>If either differential expression analysis or time-series clustering
has not been performed (depending on the selected <strong>Source of gene
list</strong>), users would be informed of the error:
<img src="images/enrich_err1.png" width="300px"></p>
<p>Otherwise, the result is shown in a table indicating the genesets’
names, number of genes in set, enrichment score (PAGE, GSEA, and
PTM-SEA), and p-values. In PTM-SEA, each reported enrichment score is
normalized by the mean of the null distribution’s enrichment score to
control for differences in signature set size. Example results are shown
in the figures below, along with explanation for each column.</p>
<p>(Pathway enrichment for Differential Expression analysis)
<img src="images/enrich_result1.png" width="520"></p>
<ul>
<li>
<strong>Name</strong>: Name of the gene set</li>
<li>
<strong>Gene Number</strong>: Number of genes in the gene set that
are also present in the result from Differential Expression
analysis.</li>
<li>
<strong>Stat</strong>: Enrichment score.</li>
<li>
<strong>p.up, p.up.adj, p.down, p.down.adj</strong>: (adjusted)
p-values estimated by the enrichment analysis.</li>
<li>
<strong>Number up</strong>: Number of genes with logFC &gt; 0</li>
<li>
<strong>Number down</strong>: Number of genes with logFC &lt; 0</li>
</ul>
<p>(Pathway enrichment for Time series clustering)
<img src="images/enrich_resultTS.png" width="440"></p>
<ul>
<li>
<strong>Gene.number</strong>: Number of genes in the gene set that
are also present in the selected cluster.</li>
<li>
<strong>Set.size</strong>: Number of genes in the gene set in the
selected database.</li>
<li>
<strong>pval, padj</strong>: p-values estimated by the Fisher’s
exact test.</li>
</ul>
<p>(Phospho-signature enrichment for Differential Expression analysis)
<img src="images/enrich_result1_Phospho.png" width="684"></p>
<ul>
<li>
<strong>Name</strong>: Name of the PTM signature set. Users should
consult the PTMsigDB website (<a href="https://proteomics.broadapps.org/ptmsigdb/" class="external-link uri">https://proteomics.broadapps.org/ptmsigdb/</a>) for more
descriptions of the names of these sets.</li>
<li>
<strong>Site.number</strong>: Number of phosphosites in the
signature set that are also in the result from Differential Expression
analysis.</li>
<li>
<strong>Number.up</strong>: Number of phosphosites with logFC &gt; 0
<strong>and</strong> are registered as up-regulated in the signature
set.</li>
<li>
<strong>Number.down</strong>: Number of phosphosites with logFC &lt;
0 <strong>and</strong> are<br>
registered as down-regulated in the signature set.</li>
<li>
<strong>Number.pSite.Db</strong>: Number of phosphosites belong to
the signature set in the selected database.</li>
<li>
<strong>Number.PTM.site.Db</strong>: Number of PTM sites belong to
the signature set in the selected database.</li>
<li>
<strong>pvalue, padj</strong>: p-value estimated by PTM-SEA.</li>
</ul>
<p>(Phospho-signature enrichment for Time series clustering)
<img src="images/enrich_result2_Phospho.png" width="294"></p>
<ul>
<li>
<strong>Name</strong>: the appendix “_upregulated” or
“_downregulated” indicates whether the phosphosites in the set are
upregulated or downregulated according to the database.</li>
<li>
<strong>Set.size</strong>: Number of phosphosites belong to the
signature set in the selected database.</li>
</ul>
<p>Clicking on a set will open a table showing the details of
genes/phosphopeptides in the set. We would like to notice that for
pathway enrichment, a gene with more than one detected phosphosite would
appear in several rows of this table, each corresponding to a
phosphosite. Nevertheless, the analysis would only consider each gene
once regardless of its number of phosphosite, thereby limiting the bias
towards genes with several sites. For phospho-signature enrichment, the
table has an additional column “PubMedID” (not shown in the figure
below), which contains the PubMed identification or hyperlink of the
source of the interaction. Multiple sources for a phosphosite are
separated by semicolons.
<img src="images/enrich_result2.png" width="318"></p>
<p>Clicking on a gene/phosphosite in this table will highlight the sets
containing it: <img src="images/enrich_result3.png" width="328"></p>
<p>The tables above can be downloaded as tsv files. Clicking on a
gene/phosphosite will also show its expression level either as a boxplot
(differential expression analysis) or a scatterplot (time-series
clustering).
<img src="images/enrich_result4.png" width="323"><img src="images/enrich_result5.png" width="326"></p>
<p>If the user has selected All time-series cluster, then a dot plot is
generated. The dot plot displays all the enriched pathways for each
cluster. Clicking on the points inside the dot plot will give the list
of genes or phosphosites associated with the particular pathway and the
cluster. Clicking on the row of the list will plot a expression level
trend for the selected genes or phosphosites.</p>
<p><img src="images/enrich_result6.png" width="906"><img src="images/enrich_result7.png" width="1001"></p>
</div>
</div>
<div class="section level2">
<h2 id="kinase-activity-inference">Kinase activity inference<a class="anchor" aria-label="anchor" href="#kinase-activity-inference"></a>
</h2>
<p>This tab performs kinase activity inference based on phosphopeptides
that are differentially expressed or in a cluster. Similar to enrichment
analysis, users would first need to perform either a differential
expression analysis or time-series clustering and select a cluster of
interest. By combining prior knowledge about known kinase-phosphosite
interactions and the data, the tab can infer the activity of the kinases
responsible for the phosphopeptides being considered. The activity is
estimated by an activity score computed with the package
<strong>decoupleR</strong> (Badia-I-Mompel et al., 2022) following the
authors’ tutorial on Kinase and Transcription Factor activity
estimation. For time-series clustering, users can also estimate how
likely the kinases are associated with phosphopeptides in the selected
cluster.</p>
<p>A network of kinase-phosphosite interactions is constructed using the
package <strong>OmnipathR</strong> (Türei et al., 2021). Users can
choose to construct this network with prior knowledge from either Homo
sapiens (taxonomy ID = 9606) or Mus musculus (taxonomy ID = 10090) by
choosing the organism in <strong>Select reference species</strong>.</p>
<p>Users need to first select whether to perform the inference on result
from either differential expression analysis or time-series clustering.
This selection affects what type of analysis can be done, as is shown in
the next two subsections.</p>
<div class="section level3">
<h3 id="differential-expression-1">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression-1"></a>
</h3>
<p>When selecting differential expression, the list of phosphopeptides
from the differential expression analysis are used for kinase activity
inference. The options to perform the inference from differential
expression analysis result are shown below:
<img src="images/kinase_settings1.png" width="300px"></p>
<ul>
<li>
<strong>Statistic used for computing the kinase score</strong>:
Whether to compute the kinase score based of the t-statistic or logFC
from the differential expression analysis.</li>
<li>
<strong>Number of top kinases in plot</strong>: How many kinases in
each direction to display in the plot.</li>
<li>
<strong>Number of permutations</strong>: How many permutations
should be done for calculating the null distribution.</li>
<li>
<strong>Highlight kinases with p-values under</strong>: Kinases
whose p-values are below this threshold will be highlighted.</li>
<li>
<strong>use FDR</strong>: Whether the FDR should be used.</li>
</ul>
<p>The output is a table showing the kinases, the activity score, and
(adjusted) p-values. Positive scores are highlighted pink while negative
ones are highlighted blue. For kinases with the most positive or
negative activity score (default 10), their scores are also plotted in a
horizontal barplot. Kinases whose p-values under the threshold are
colored red. <img src="images/kinase_result1.png" width="462"></p>
<p><strong>Interpreting the result</strong>: The kinase activity is
computed from comparing two conditions, hence it is important to take
into account the conditions when interpreting the result. A positive
score means that the kinase is more active in the selected condition
compared to the reference one, while a negative score means that the
kinase is more active in the reference condition.</p>
<p>Selecting a kinase will open a table showing the phosphosites
targeted by the kinase. The information shown in the table is from the
differential expression analysis: log2FC, t-statistic, peptide sequence,
and (adjusted) p-values. We would like to notice that these p-values are
from the differential expression analysis and should not be confused
with those derived from the kinase activity inference.
<img src="images/kinase_result2.png" width="438"></p>
</div>
<div class="section level3">
<h3 id="time-series-clustering-1">Time-series clustering<a class="anchor" aria-label="anchor" href="#time-series-clustering-1"></a>
</h3>
<p>When selecting time-series clustering, the list of phosphopeptides in
the selected cluster (in the <strong>Time series clustering</strong>
tab) is used to either infer the kinase activity score or how likely the
kinases are associated with the cluster. These analyses will be
explained in more details in the following 2 subsections.</p>
<div class="section level4">
<h4 id="estimating-kinase-activity-score">Estimating kinase activity score<a class="anchor" aria-label="anchor" href="#estimating-kinase-activity-score"></a>
</h4>
<p>For this part, users would need to define the number of permutations
for calculating the null distribution.
<img src="images/kinase_settings2.png" width="300px"></p>
<p>Then the option to perform clustering specified in the <strong>Time
series clustering</strong> tab affects the analysis being done in this
part. If the option <strong>expression</strong> was chosen, the
fold-change in expression levels for consecutive time points would be
used for the inference. Thus, the result would reflect how the kinase
activity change over time. Below is an example of the result when
performing the inference with 4 time points:
<img src="images/kinase_result3.png" width="459"></p>
<ul>
<li>The table shows the kinases along with its activity score and
(adjusted) p-values. The timepoint column shows which two time points
were considered when estimating the score. Activity scores are labeled
pink if they are positive and blue if negative.</li>
<li>The kinase activity scores are plotted in a heatmap and those whose
p-values below the threshold are highlighted by asterisk.</li>
</ul>
<p>Clicking on a kinase will open a table showing the phosphosites
targeted by the kinase:
<img src="images/kinase_result4.png" width="328"></p>
<p>If the <strong>logFC</strong> or <strong>two-condition
expression</strong> option was chosen for time-series clustering, the
fold-change in expression levels between the two treatments would be
used for the inference. Thus, the result would reflect how the kinase
activity differ between two treatments at each time point. The layout of
the result is similar to that of the above-mentioned
<strong>expression</strong> option:
<img src="images/kinase_result5.png" width="458"></p>
<p>If no kinase was found (wrong reference species or no overlap with
prior knowledge), users will be informed of the error:
<img src="images/kinase_err1.png" width="278"></p>
</div>
<div class="section level4">
<h4 id="estimating-kinase-association">Estimating kinase association<a class="anchor" aria-label="anchor" href="#estimating-kinase-association"></a>
</h4>
<p>In addition to kinase activity, users can choose to determine whether
the kinases are associated with the selected cluster, using either the
Fisher’s exact test or the Fast Gene Set Enrichment Analysis algorithm
(FGSEA). In FGSEA, the probability of each site to be in the cluster is
used to rank the sites. Performing FGSEA is done by the decoupleR
package (Badia-I-Mompel et al., 2022) and requires users to set the
number of permutations for calculating the null distribution. Both
methods use the kinase-phosphosite interaction network as their
reference. Unlike the above part on estimating kinase activity, the
choice of method for time series clustering in the <strong>Time series
clustering</strong> tab does not affect the method in this part.
<img src="images/kinase_settings3.png" width="300px"></p>
<p>The result for the Fisher’s exact test is displayed in a table.
Kinases whose (adjusted) p-values higher than the p-value threshold are
excluded. <img src="images/kinase_result6.png" width="250"></p>
<ul>
<li>
<strong>number.pSite.in.cluster</strong>: Number of phosphosites
targeted by the kinase and are present in the cluster.</li>
<li>
<strong>number.pSite.by.kinase</strong>: Number of all phosphosites
known to be targeted by the kinase, as derived from the above-mentioned
kinase-phosphosite network.</li>
<li>
<strong>p-value</strong>: p-value from the test.</li>
<li>
<strong>padj</strong>: adjusted p-value (FDR).</li>
</ul>
<p>The result for the FGSEA is also displayed in a table showing the
enrichment score.
<img src="images/kinase_result7.png" width="300px"></p>
<p>Clicking a kinase will open a table showing the phosphosites targeted
by the kinase in the cluster.
<img src="images/kinase_result8.png" width="327"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="log-info">Log Info<a class="anchor" aria-label="anchor" href="#log-info"></a>
</h2>
<p>This tab displays the information about the user inputs. All the
settings selected by the user are displayed in a table and can be
downloaded as a TSV file. The table contains the following columns:</p>
<ul>
<li>
<strong>name</strong>: name of the input (usually the widget ID used
in the shiny app script).</li>
<li>
<strong>value</strong>: the value associated with the name. This is
the actual user input.</li>
<li>
<strong>panel</strong>: the tab panel in the shiny app to which this
value belongs.</li>
<li>
<strong>description</strong>: a short description about that
particular setting.</li>
</ul>
<p><img src="images/log_info.png" width="862"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] BiocStyle_2.36.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] digest_0.6.37       desc_1.4.3          R6_2.6.1           </span></span>
<span><span class="co">##  [4] bookdown_0.43       fastmap_1.2.0       xfun_0.52          </span></span>
<span><span class="co">##  [7] cachem_1.1.0        knitr_1.50          htmltools_0.5.8.1  </span></span>
<span><span class="co">## [10] png_0.1-8           rmarkdown_2.29      lifecycle_1.0.4    </span></span>
<span><span class="co">## [13] cli_3.6.5           sass_0.4.10         pkgdown_2.1.3      </span></span>
<span><span class="co">## [16] textshaping_1.0.1   jquerylib_0.1.4     systemfonts_1.2.3  </span></span>
<span><span class="co">## [19] compiler_4.5.1      tools_4.5.1         ragg_1.4.0         </span></span>
<span><span class="co">## [22] evaluate_1.0.4      bslib_0.9.0         yaml_2.3.10        </span></span>
<span><span class="co">## [25] BiocManager_1.30.26 jsonlite_2.0.0      rlang_1.1.6        </span></span>
<span><span class="co">## [28] fs_1.6.6            htmlwidgets_1.6.4</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Badia-I-Mompel, P., Santiago, J. V., Braunger, J., Geiß, C.,
Dimitrov, D., Müller‐Dott, S., Tauš, P., Dugourd, A., Holland, C. H.,
Flores, R. O. R., &amp; Sáez-Rodríguez, J. (2022). decoupleR: ensemble
of computational methods to infer biological activities from omics data.
Bioinformatics Advances, 2(1). <a href="https://doi.org/10.1093/bioadv/vbac016" class="external-link uri">https://doi.org/10.1093/bioadv/vbac016</a></p>
<p>Kim, SY., Volsky, D.J. PAGE: Parametric Analysis of Gene Set
Enrichment. BMC Bioinformatics 6, 144 (2005). <a href="https://doi.org/10.1186/1471-2105-6-144" class="external-link uri">https://doi.org/10.1186/1471-2105-6-144</a></p>
<p>Krug, K., Mertins, P., Zhang, B., Hornbeck, P., Raju, R., Ahmad, R.,
Szucs, M. J., Mundt, F., Forestier, D., Jané‐Valbuena, J., Keshishian,
H., Gillette, M. A., Tamayo, P., Mesirov, J. P., Jaffe, J. D., Carr, S.
A., &amp; Mani, D. R. (2019). A curated resource for
phosphosite-specific signature analysis. Molecular &amp; Cellular
Proteomics, 18(3), 576–593. <a href="https://doi.org/10.1074/mcp.tir118.000943" class="external-link uri">https://doi.org/10.1074/mcp.tir118.000943</a></p>
<p>Liberzon, A., Birger, C., Thorvaldsdóttir, H., Ghandi, M., Mesirov,
J. P., &amp; Tamayo, P. (2015). The Molecular Signatures Database
(MSigDB) hallmark gene set collection. Cell systems, 1(6), 417–425. <a href="https://doi.org/10.1016/j.cels.2015.12.004" class="external-link uri">https://doi.org/10.1016/j.cels.2015.12.004</a></p>
<p>Liberzon, A., Subramanian, A., Pinchback, R. M., Thorvaldsdóttir, H.,
Tamayo, P., &amp; Mesirov, J. P. (2011). Molecular signatures database
(MSigDB) 3.0. Bioinformatics, 27(12), 1739–1740. <a href="https://doi.org/10.1093/bioinformatics/btr260" class="external-link uri">https://doi.org/10.1093/bioinformatics/btr260</a></p>
<p>Subramanian, A., Tamayo, P., Mootha, V. K., Mukherjee, S., Ebert, B.
L., Gillette, M. A., Paulovich, A. G., Pomeroy, S. L., Golub, T. R.,
Lander, E. S., &amp; Mesirov, J. P. (2005). Gene set enrichment
analysis: A knowledge-based approach for interpreting genome-wide
expression profiles. Proceedings of the National Academy of Sciences of
the United States of America, 102(43), 15545–15550. <a href="https://doi.org/10.1073/pnas.0506580102" class="external-link uri">https://doi.org/10.1073/pnas.0506580102</a></p>
<p>Türei, D., Valdeolivas, A., Gul, L., Palacio-Escat, N., Klein, M.,
Ivanova, O., Ölbei, M., Gábor, A., Theis, F. J., Módos, D., Korcsmáros,
T., &amp; Sáez-Rodríguez, J. (2021). Integrated intra‐ and intercellular
signaling knowledge for multicellular omics analysis. Molecular Systems
Biology, 17(3). <a href="https://doi.org/10.15252/msb.20209923" class="external-link uri">https://doi.org/10.15252/msb.20209923</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Shubham Agrawal, Junyan Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
